{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Rosalind Exercises\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 1 - Counting DNA Nucleotides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[233, 234, 206, 229]\n"
     ]
    }
   ],
   "source": [
    "seq = \"\"\n",
    "\n",
    "with open(\"exercises/rosalind_dna (3).txt\") as file_in:\n",
    "    for line in file_in:\n",
    "        seq += line\n",
    "\n",
    "count = [seq.count('A'),seq.count('C'),seq.count('G'),seq.count('T')]\n",
    "print(count)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 2 - Transcribing DNA into RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCCGGCUGUAAGCGAAAGCGCCGACUACAAAUGAUGGUCGACUCACUUAUACGCGAGAAAAGUUUAUAUGACUAGGGGGGGUGACCAGACACAUGAACCCGUGUCCAAGGUUUACGGAAUCUAAGUUGUGGCCCGAUAUGAGUAAGUAUAGCUCACCCCAAACCAUUUGUGCGCUGAUCUGGGCGAGGAGUGGAGGGUGCCCUCUCGAGAGCCUACCGAUAGUCCACGGAAGCGAUGAUGUUAUACGGUGCACGGCAAACUAACCCUGCUGGAGGAAAGGUUCCGCCAGGGUCACUUCCACACCCGAUCGUAGCAUGGGGACGACGAUUAGAUGGAGAAUCCGGCGCGUAAACUGGGCACAUUGACUUCAUCAGGUGCAUCGCUAAUCAUUCCAAUCAUAUUUAGUCCUCCUCAAGACAAGAAUACCGCCAUUACUACCCAACUACCUAAGUACAGAUACGGAAGGGCAUGACAAUCCCAGUGUGUACAUCUCCCGGAAGAUCCUUGCCUCUUACCCAGUGUAACGGCCAUAUCAGUCCAUUCAGACCCUGAAUAAGAAGAACAAGCACCCACUCCCCUAGGGCCACCCCUGCCGCAUUACAGUUGUGUACUAUUGUCACAGUAGGAAUCUAAAGUGAUAGCAGGAUCGAUGUAUCGAAUGUACUAGGCCAGGGACGACCGGCAUGUAAAUCUUCGAAGGGUGAGUAGCAAGGGGAAAAAUGUGCGUUAACUGCCAGACGCCCUCUCUAAUGAGACAAAUCAGUUUCGUCCGUGUGUCAGAAAAUGGAAAAUCUGCCGGUACUGCGAACAUGAGACCCGCCCUCCCAACACGCCGUAAAUGGAUUAAUCAGCGGCCAGGCGGCUAUGGGAGAUCUCAGGACUCUUCAGAGUUUGCAGCCCUCAGACUACGUCAUGGUUACUGCGCGAGUAAUAGAGUUG\n",
      "\n"
     ]
    }
   ],
   "source": [
    "seq = \"\"\n",
    "\n",
    "with open(\"exercises/rosalind_rna.txt\") as file_in:\n",
    "    for line in file_in:\n",
    "        seq += line\n",
    "\n",
    "rnaseq = \"\"\n",
    "for nuc in seq:\n",
    "    if nuc == 'T':\n",
    "        rnaseq += 'U'\n",
    "    else:\n",
    "        rnaseq += nuc\n",
    "\n",
    "print(rnaseq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 3 - Complementing a Strand of DNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TACGCGGTCTGGTGTGGGTAACTCATAGTCATTTTCAATCAGCTTGCAATAGCATACCCTCCGACCGCAGATAGGGATCGCGGATGCGTATGAAGAAATCACATGGCCACTGGTAGTAGGCGCAGTTGGGAAAGGAATAATGAATTAAGACCGGCTAACACCAGCTTATGATCCTGTGCTGTTATGGGTGAACAAGTAGGGAGGCTCTAACCGCGTTGATTCACGACTATATCATTTCATCTGGACACTAGGTGCGTGACCGTCTGCAGCAAAGTAATTGCGTGCGTGCTTGTGCTAGGTATATGCGTTGTATACCGACGCGGTAAGATTGCCTGTCAGACCTGGAGCCTAAGACTAGACCCAGGCGTCAACCGTAAAGGAGTTTGATAGACATAGGCTTCCGTTAACTACTAACCCTGTAAGAGAGCCCAAATAATGTAAGGTAGCTTGGCTTTTTTAGAGTAGCCACTACACATCTTGAAGGACAGATAATTACTATCCGGGAAGGCTATTTGTAAAGACATATGCTTGGTTTAAACATGAAGATTGATCTCGTAACTCTACGATACGATATAAGGGGGCTTGGGCCTGACTCGTGAATAACTGAAGTCGTTTCGTCCTGGTTGGGTTTAACTCTCAAGGAGGGAACATTTGAGAACTTCGCACATCGCTTGAAATCGTGCGACCGATCCTAGGGGCACAAGTTATGTGCGATGTCATTGCTATCGCGGTGGTCCAGCCCCCGCCGCGTGGTTTCGTGATGTGGTATGAGTAAGGAAGACAGTTGCGCGGCCAGCTGGTGATGAGCTATTGGCTAAACAGATGATAATAATGGCCGTGTATGGACCTTACGCACGGCAATCACCCCTGTACGCCTCTTGGGCTCCCGTTTTAGGTGTG\n"
     ]
    }
   ],
   "source": [
    "seq = \"\"\n",
    "\n",
    "with open(\"exercises/rosalind_revc.txt\") as file_in:\n",
    "    for line in file_in:\n",
    "        seq += line\n",
    "        seq = seq.strip()\n",
    "dictionary = {\"A\":\"T\", \"T\":\"A\", \"C\":\"G\", \"G\":\"C\"}\n",
    "\n",
    "seq2 = \"\"\n",
    "for nuc in seq:\n",
    "    seq2 += dictionary[nuc]\n",
    "\n",
    "seq2 = seq2[::-1]\n",
    "print(seq2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The slice operator [n:m] returns the part of the string from the n’th character to the m’th character, including the first but excluding the last. In other words, start with the character at index n and go up to but do not include the character at index m. This behavior may seem counter-intuitive but if you recall the range function, it did not include its end point either.\n",
    "\n",
    "If you omit the first index (before the colon), the slice starts at the beginning of the string. If you omit the second index, the slice goes to the end of the string.\n",
    "\n",
    "There is no Index Out Of Range exception for a slice. A slice is forgiving and shifts any offending index to something legal.\n",
    "\n",
    "https://stackoverflow.com/questions/509211/understanding-slice-notation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 4 - Rabbits and Recurrence Relations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1, 1, 5, 9, 29, 65, 181, 441, 1165, 2929, 7589, 19305, 49661, 126881, 325525, 833049, 2135149, 5467345, 14007941, 35877321, 91909085, 235418369, 603054709, 1544728185, 3956947021, 10135859761, 25963647845, 66507086889, 170361678269, 436390025825, 1117836738901, 2863396842201, 7334743797805, 18788331166609, 48127306357829]\n"
     ]
    }
   ],
   "source": [
    "#From rosalind_fib.txt\n",
    "\n",
    "n = 35\n",
    "k = 4 #number of pairs produced per mature pair\n",
    "pop = [1] #starting pair; each pair will take 1 month to mature and will only produce offspring by 2nd month of life\n",
    "for i in range(0, n-1):\n",
    "    pop.append(pop[i]) #Initialise the list with a number of spaces equivalent to n months\n",
    "for i in range(2, n):\n",
    "    pop[i]=pop[i-2]*k+pop[i-1] #Recurrence relation where number of pairs alive = n(pairs alive last month) + k*n(mature pairs)\n",
    "print(pop)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 5 - Translating RNA into Protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MHKAKTTIAQCSQCCCVTFMGKRHRFNAKSRLLAHTIIIIIWKYASCIPLVADSGRGLLRPLSKWRQMACSLLHLDYERHYLSKRNLRMRAPIAGEKSPRTQGYVLWGVNRYGSPRDCLRSQVPASEVHFYNSARLLPNLAKAFGVVLHKALRKIHPRQGSFGYMVGCVYRLLQATNWAPCRDISGCDRRCRKRSPVLCLRRHRYQQTATLQSLYQYKEHPVGMHFVMHSNQWIYAIFNIGCYKYGLGTIGTSLGDVPTTPLPSFPWNSGVFWLSPRPGIAERNHCWGGAKSRRVQDKYPTCLQHGHDSSGGIAPVDTLHRTHRRGGFQKDASALPKYACDWFVRPGQSPASSSIVLAGGSFSFRLLVDIFEDIKRKLFKIGGLGDSPEPLCRGLLFFGCSPKAQSLDYPAVTRYTGVECTSVGRRPLTCHELTLLWNMQVDLIELLRPKLASLTTVRVCYRDDAVTLGLVRTSESKQRRYERSKEGGRQLWVGTGQRILTTGRGFFDSARLALGIFTTTTIPQTPKPPSDRACRDNLSVTSADRLLCFRITPAYNDPSLLAGLSPLTVDVLKYTLSGDTRRPLPLRHVGPTMLNLVFAYITQSFLISLFRTLKVPSSLHAERRPRIARECVVSRSESASFTKTLCPGSRILPLPWLTIAPQAKPRLPRSVLSSLEHFGFTNNRANIVSYEFTHGFCRTHERPLLVCTLTTILGSNFARTRIFCASRTIARYRSDLRTDNQSTVSLPRGLRSDGGDLRARTRRRQNSHDDDVLRLGYLQVKNMMSQPGAPCGAYITPGTRVAQQSIQALQVNLARHGGPASRGRHTEVASFALHVGRSILSGVRHWRWPSELQVEVSQPFMTKMNTERKQAPMISERSQFAYSYTKIRGLVLFGPRAHRQILLSVYTSSQQDCFIVSLLFIVHVKFMAYTRLRCYMSLFHLRDIKKVSSARPPSRHLSHEPQKPLCDPSICRKIPNYAGAVAADPVQIFPSEALPTSRISRACDRAGLHTAGDHGLHRNGVRMPTDSFAYGSGLVLRRETALVARPVCESYENPSIHRANKGLSQFAIVQSHRSSEPLSLGSLLRFVHPLCVTARSQVPALPPYQLARPLSCNNGENWDYDSWCESSRRSLRRVSSTDLGKLCVRALDCEVSTYCIIRVTVLDASQCRHCPLRYCDTQNSTYHRCMPLNVPVIRLGGFGTINSPRLKFEEYMCAVVAPGACWKTMFADSCTIECIIGNSQHLSSGSIERDSRSLPQILSVPLKSVLEPVISSSPIKPIIFPSSLTGRPNSGSNEIFTHCQGVSYPWWDVEYGTRAVLSLIANLRQSNRGVATSVIAIYRAAVLGNRASMSRKFSHHVCFEGVSHGDDPSAHDELHLLFYRRASRHVFLPYDTDALCVRQHRVAYRRAVYSTAWACILLLALELNQIYLLGHRNRGDRSGCNVKHATKIQKGGTLAAIVRLRSDFIIRTGYISGRGMVDRLHYSASGSGVSSGTRMPKGANKAIPNSDHDIVDRMIDRRIRIVSDAVHYRAHQKVSNDHDCLQKEGWEYLACFLREGLKFAEGLHVMTFSRLQDSFCQIVAIRPLSVLHGLDAAARPLWRHQALRYRAAWQIHTYQLSITHISMRQLSRSWGGPTRPSIEAKRAFLTDTARDQFIKQSGFHVEDTRASGSVSIIVFPRQLGCLRKSSRLHLSKCFLNSCFPMMANGVGSCSGMDASRGGCRSAFLYGSACTRLYQQFYLVIGIWVTPPIEPMLLTQYAPACGVTTGQPGSIELSVGRLHTNCDTDLSIERGRVGLQEPRAFRSVLKKVGILCIEMYRRVPPVKDQSVRRSIDRVGQHVTHRALSGCPAIVISTRGITGCDPYYGRNLQFGHGSVDRSVLYASGLFSDYISIWFRSFNLSDTAWNSEMHFALAATCPGFISTRQRISVITLGYSSSLRQRYSSQEPFCYWGSIGSRLPFFHSHGNHAICVGKKVKTFRISPYLDWVPCENNVIPFRKPEQLECSYSGNPAAFDPVYKLADVEPIRGEAEPIRDEVLRQPPVSAHSVYCERARQLRNVTLKENEDTEGFQIYCVVIIAQVPTAYFPSFRRIGREGGDRLLSPHVGKHPPVRSRTTFNQNNLLLVATPFPADDLIARRVSVTFMMLYRPIGSCKVQCVLQVARNAAEVTVWRSIRASRFALPLPVLTFANPDYTSAFLQRSFAPSDKRNRPLYFCERVYESSHGAILTSHYRCAHIRKKWKSHWLGYSFTEVLCQSISRNTREKLMLMISKGRLWSDTLRHLRHSPYKEFHLNGFLQQGGQSTNSVYAEYCLKARDLAMQLLVRTYESKYGVSSSYLGIKRDGPVLRLREGTLRPHLPFRKSIWRRKAYVGSGTSSTDKVLLIYNPDLLGDVHGEALAFQVYCSASRNSVSPVGMLLTFAPSLLAGLSPNTYSGIVGGSTSISRLTLPKQTRGSRDSGGPCKASPRVPAATLDEDTDSSRMSYKTGAFSIYASRRSSSQKKLARTLASVRSWCLLRECNDEALNTTPISALEAAWGRQPAMLYWSDDAKSPNLGSPFSIVGVSWLTPNSEQVALDQLGFNRAYSRGSILGGSLPGAHEALPVHTLSERRSLESLQRFSPVAESPEVRSKRQSADSMQSGISSLCFSSTCRAAPCVRAGAILSYLPDTVSGRVLLTDCCQSKPELTSGTVLKRTTSSADLTIGGLHPTNNVKLLISNRSTYNKGSDVTKSIGLGHHIMCEHVPESGPWPEPSLPNISEFFGHHSHGKLRRVSVLISLGYERDLFCAIKVQNAVAQASGRWAPLEKLSYNQSWLGTGSIVRAGRPLGFYFRQGSCLLMLLVCTVGLSAVSIVVHGLVTTRGQTALLKNDRHGLLPSIRTLHLSQYHYRVLVEGRLVSLNPVPRRVRRKNPNSSCGSTQDQSYSSATHTTIAMDFTGLVVRARPCSRIIWSQSAGTTAHDSGKGGDAINRVPIARNFTPRVARGANDSIPRRRQQMVPQLVVSTSHAVVNVCLRDLMATNLNPPRTFSTSSGHYEPYSLNVIIQVWMVPEVFGRTGPRIRRQLSKSIPQGRMTRLY\n"
     ]
    }
   ],
   "source": [
    "seq = \"\"\n",
    "\n",
    "with open(\"exercises/rosalind_prot (1).txt\") as file_in:\n",
    "    for line in file_in:\n",
    "        seq += line\n",
    "        seq = seq.strip()\n",
    "dictionary = {\"UUU\" : \"F\",\"CUU\" : \"L\",\"AUU\" : \"I\",\"GUU\" : \"V\",\"UUC\" : \"F\",\"CUC\" : \"L\",\"AUC\" : \"I\",\"GUC\" : \"V\",\"UUA\" : \"L\",\"CUA\" : \"L\",\n",
    "              \"AUA\" : \"I\",\"GUA\" : \"V\",\"UUG\" : \"L\",\"CUG\" : \"L\",\"AUG\" : \"M\",\"GUG\" : \"V\",\"UCU\" : \"S\",\"CCU\" : \"P\",\"ACU\" : \"T\",\"GCU\" : \"A\",\n",
    "              \"UCC\" : \"S\",\"CCC\" : \"P\",\"ACC\" : \"T\",\"GCC\" : \"A\",\"UCA\" : \"S\",\"CCA\" : \"P\",\"ACA\" : \"T\",\"GCA\" : \"A\",\"UCG\" : \"S\",\"CCG\" : \"P\",\n",
    "              \"ACG\" : \"T\",\"GCG\" : \"A\",\"UAU\" : \"Y\",\"CAU\" : \"H\",\"AAU\" : \"N\",\"GAU\" : \"D\",\"UAC\" : \"Y\",\"CAC\" : \"H\",\"AAC\" : \"N\",\"GAC\" : \"D\",\n",
    "              \"UAA\" : \"\",\"CAA\" : \"Q\",\"AAA\" : \"K\", \"GAA\" : \"E\",\"UAG\" : \"\",\"CAG\" : \"Q\",\"AAG\" : \"K\",\"GAG\" : \"E\",\"UGU\" : \"C\",\"CGU\" : \"R\",\n",
    "              \"AGU\" : \"S\",\"GGU\" : \"G\",\"UGC\" : \"C\",\"CGC\" : \"R\",\"AGC\" : \"S\",\"GGC\" : \"G\",\"UGA\" : \"\",\"CGA\" : \"R\",\"AGA\" : \"R\",\"GGA\" : \"G\",\n",
    "              \"UGG\" : \"W\",\"CGG\" : \"R\",\"AGG\" : \"R\",\"GGG\" : \"G\"}\n",
    "prot = \"\"\n",
    "for i in range(0, len(seq),3):\n",
    "    prot += dictionary[seq[i:i+3]]\n",
    "print(prot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 6 - Mendel's First Law"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.7793\n"
     ]
    }
   ],
   "source": [
    "#From rosalind_iprb(1).txt\n",
    "\n",
    "k = 30 #AA\n",
    "m = 17 #Aa\n",
    "n = 26 #aa\n",
    "total = k+m+n\n",
    "\n",
    "prob1 = (n/total)*((n-1)/(total-1)) #aa x aa\n",
    "prob2 = (m/total)*(n/(total-1)) #Aa x aa; we need to account for aa x Aa as well!\n",
    "prob3 = (m/total)*((m-1)/(total-1))*0.25 #Aa x Aa\n",
    "                      \n",
    "recessiveprob = prob1+prob2+prob3\n",
    "domprob = round(1-recessiveprob, 5)\n",
    "\n",
    "print(domprob)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 7 - Computing GC Content"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This one was especially difficult due to the for loop - as I had used labels as markers to chop and append the sequences before them, I could not print the last sequence. The solution seems obvious in retrospect - as the for loop only runs for the number of lines in a file, the last sequence is still stored in the seq variable, and can be retrieved with a simple append command after the for loop."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rosalind_9195\n",
      "51.373\n",
      "Rosalind_2025\n",
      "51.83432\n",
      "Rosalind_9187\n",
      "47.59036\n",
      "Rosalind_0633\n",
      "52.81863\n",
      "Rosalind_9029\n",
      "49.71297\n"
     ]
    }
   ],
   "source": [
    "seq =''\n",
    "stuff = [] #Create an empty list to store values - both labels and sequences \n",
    "with open(\"exercises/rosalind_gc (1).txt\") as file_in:\n",
    "    for line in file_in:\n",
    "        if line[0] == \">\":\n",
    "            line = line.lstrip('>')\n",
    "            line = line.rstrip('\\n')\n",
    "            stuff.append(seq)\n",
    "            stuff.append(line)\n",
    "            seq = \"\"\n",
    "        else:\n",
    "            line = line.rstrip('\\n')\n",
    "            seq += line\n",
    "    stuff.append(seq) #When it runs out of lines, it'll append the last sequence\n",
    "\n",
    "for i in range(2, len(stuff), 2):\n",
    "    GC = 0\n",
    "    for nuc in stuff[i]:\n",
    "        if nuc == \"C\" or nuc == \"G\":\n",
    "            GC += 1\n",
    "    print(stuff[i-1]) #Print label\n",
    "    print(round((GC/len(stuff[i]))*100,5)) #Print sequence GC content"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 8 - Finding a Motif in DNA "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that string indices begin with 0, but the outputs begin with 1. Use the sample data to debug."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GCATTTAACAATGCATAATGCATAATGCATCGAATGCATAATGCATAATGCATATAATGCATTAATGCATCTAATGCATAGGGGAATGCATAAATGCATAATGCATAATGCATGATTGAATGCATAATGCATAATGCATAAGCAATGCATGTAAATGCATTCAATGCATTAAATGCATGCAAATGCATGAATGCATTGCCTAATGCATGTTCTAGCTGAATGCATAATGCATAATGCATCAATGCATAATGCATAATGCATTAATGCATAATGCATAATGCATAATGCATCGAAATGCATAGTGTAATGCATGCCAATGCATAATGCATTAAATGCATAATGCATAAGAATGCATAATGCATAATGCATGAATGCATAATGCATGCAAATGCATCCTAATGCATAAATGCATAATGCATCAGAATGCATAGTTAATGCATTAATGCATAATGCATCTCAATGCATAATGCATTGCTTTAATGCATAATGCATAATGCATTTGGAGATAATGCATATTTGTTAATGCATAAAATGCATTCTGAATGCATCTAAATGCATATAATGCATACTAATGCATCAATGCATAATGCATTAATGCATAATGCATTAGGACTGGTCAATGCATCAATGCATTTAATGCATAAATGCATAATGCATTAATGCATTAATGCATGGAATGCATGAAATGCATGGACAATGCATTCAATGCATAATGCATAATGCATAAATGCATGAAATGCATACATAAATGCATCGAATGCATAATGCATTGCATACAATGCATAAACACTAATGCATGTTTAATGCATGCAATGCATGCTCGAAGCAATGCATTAATGCATAATGCATGCTAATGCATAAATGCATCTTTGATCGAATGCATAAATGCATGCGGGAATGCATATATACAAATGCATGGAAATGCATAATGCATTTCCAATGCATGCATGACAGAACATCCTT', 'AATGCATAA']\n",
      "10 17 33 40 85 93 100 119 126 133 219 226 241 248 263 270 277 316 332 339 349 356 371 398 406 442 459 479 486 522 579 594 636 644 705 712 719 757 778 836 853 877 921 "
     ]
    }
   ],
   "source": [
    "linelist = []\n",
    "with open(\"exercises/rosalind_subs (1).txt\") as file_in:\n",
    "    for line in file_in:\n",
    "        line = line.rstrip(\"\\n\")\n",
    "        linelist.append(line)\n",
    "\n",
    "print(linelist)\n",
    "seq = linelist[0]\n",
    "motif = linelist[1]\n",
    "j = len(motif)\n",
    "i = 0\n",
    "\n",
    "for i in range(0, len(seq)):\n",
    "    if seq[i:i+j] == motif:\n",
    "        print(i+1, end = \" \")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 9 - Counting Point Mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "434\n"
     ]
    }
   ],
   "source": [
    "linelist = []\n",
    "with open(\"exercises/rosalind_hamm.txt\") as file_in:\n",
    "    for line in file_in:\n",
    "        line = line.rstrip(\"\\n\")\n",
    "        linelist.append(line)\n",
    "        \n",
    "seq1 = linelist[0]\n",
    "seq2 = linelist[1]\n",
    "H = 0\n",
    "\n",
    "for i in range(0, len(seq1)):\n",
    "    if seq1[i] != seq2[i]:\n",
    "        H+=1\n",
    "print(H)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 10 - Calculating Expected Offspring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "149982.0\n"
     ]
    }
   ],
   "source": [
    "linelist = []\n",
    "with open(\"exercises/rosalind_iev.txt\") as file_in:\n",
    "    for line in file_in:\n",
    "        line = line.rstrip(\"\\n\")\n",
    "        linelist.append(line)\n",
    "\n",
    "seq1 = linelist[0].split()\n",
    "\n",
    "for i in range(0, len(seq1)): #Coerce all floats to integers since data given is all integers anyway\n",
    "    seq1[i] = int(seq1[i])\n",
    "\n",
    "Case1 = seq1[0]*2 #AA-AA\n",
    "Case2 = seq1[1]*2 #AA-Aa\n",
    "Case3 = seq1[2]*2 #AA-aa\n",
    "Case4 = seq1[3]*2*0.75 #Aa-Aa\n",
    "Case5 = seq1[4]*2*0.5 #Aa-aa\n",
    "Case6 = seq1[5]*0 #aa-aa\n",
    "\n",
    "Expected = Case1+Case2+Case3+Case4+Case5+Case6\n",
    "print(Expected)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 11 - Mortal Fibonacci Rabbits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "676172117379838466\n"
     ]
    }
   ],
   "source": [
    "#From rosalind_fibd (1).txt\n",
    "\n",
    "#Here, each mature pair only gives rise to a pair of offspring.\n",
    "\n",
    "n = 87 #number of months\n",
    "m = 18 #lifespan per pair - they only live for m months.\n",
    "ages = [0]*m #Create a cohort table to follow the population through m months\n",
    "ages[-1] = 1 #We have a pair with m months left to live\n",
    "\n",
    "for i in range (1,n):\n",
    "    new_rabbits = sum(ages[0:-1]) #We want to take a snapshot of all mature rabbits, i.e. all except newborns with m months left to live\n",
    "    ages[:-1] = ages[1:] #We want to shift all rabbits left as they age - ages[0] = ages[1]...ages[-2] = ages[-1]\n",
    "    ages[-1] = new_rabbits #Finally, we add the newborns to the end - we didn't do this first so we wouldn't overwrite any data\n",
    "\n",
    "print(sum(ages)) #The total number of rabbits is simply the sum of rabbits of all ages"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For interest, the recurrence relation is Fn = F(n-1) + F(n-2) - F(n-(m+1)); however, you need to specify the edge case when n = m, because it will go out of bounds and give funny results.\n",
    "\n",
    "Generally what is done is that three cases are specified, after the sequence is seeded with 1,1 and months = 2. While months is less than n:\n",
    "- Normal Fibonacci when months is less than m - no one has died yet\n",
    "- Subtract 1 when months = m and months = m + 1 to indicate the death of the first pair\n",
    "- Subsequently, when months is between m+1 and n, use the modified recurrence relation above\n",
    "\n",
    "and then months += 1, return last item of list. \n",
    "\n",
    "We need the months to be strictly less than n because this is computing the next month's population during the previous month. So if n = 62, it would return the population at n = 62 when months = 61, because the first input is fed into the formula when months = 2 to give us the population for n = 3. Only then is months updated post hoc."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 12 - Inferring mRNA from Protein "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "345728\n"
     ]
    }
   ],
   "source": [
    "linelist = []\n",
    "with open(\"exercises/rosalind_mrna.txt\") as file_in:\n",
    "    for line in file_in:\n",
    "        line = line.rstrip(\"\\n\")\n",
    "        linelist.append(line)\n",
    "        \n",
    "seq = linelist[0]\n",
    "comb = 1\n",
    "\n",
    "dictionary = {\"A\":4, \"R\":6, \"N\":2, \"D\":2, \"C\":2, \"Q\":2, \"E\":2, \"G\":4, \"H\":2, \"I\":3, \"L\":6, \"K\":2, \"M\":1, \"F\":2, \n",
    "           \"P\":4, \"S\":6, \"T\":4, \"W\":1, \"Y\":2, \"V\":4}\n",
    "\n",
    "for amino in seq:\n",
    "    comb = comb*dictionary[amino]\n",
    "\n",
    "comb = comb*3\n",
    "output = comb%1000000\n",
    "print(output)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 13 - Independent Alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.6142569731400068\n"
     ]
    }
   ],
   "source": [
    "#from rosalind_lia (1).txt\n",
    "from math import comb\n",
    "\n",
    "k = 7 #kth generation\n",
    "n = 31 #at least this number of AaBb individuals\n",
    "\n",
    "#I did the allele table calculations and found that irrespective of genotype, if you are crossed with AaBb, you get AaBb 1/4 of the time.\n",
    "\n",
    "offspring = pow(2,k)\n",
    "prob = [pow(0.75, offspring)] #probability of 0 individuals being AaBb - 0.75^(2^k)\n",
    "\n",
    "for i in range(1,n):\n",
    "    NewCase = pow(0.75, offspring-i)*pow(0.25, i)*math.comb(offspring, i)\n",
    "    prob.append(NewCase) #All cases for 1, 2...n-1 individuals being AaBb, accounting for all possible combinations\n",
    "    #Consider that if n = 1, there are 2^k different offspring to be that sole AaBb - you must account for all possibilities.\n",
    "\n",
    "hetprob = 1-sum(prob) #The idea here is that we do 1 - P(AaBb<n) = P(AaBb>=n)\n",
    "\n",
    "print(hetprob)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 14 - Calculating Protein Mass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "99429.752\n"
     ]
    }
   ],
   "source": [
    "linelist = []\n",
    "with open(\"exercises/rosalind_prtm.txt\") as file_in:\n",
    "    for line in file_in:\n",
    "        line = line.rstrip(\"\\n\")\n",
    "        linelist.append(line)\n",
    "        \n",
    "seq = linelist[0]       \n",
    "\n",
    "dictionary = {\"A\":71.03711,\"C\":103.00919,\"D\":115.02694,\"E\":129.04259,\"F\":147.06841,\"G\":57.02146,\"H\":137.05891,\n",
    "              \"I\":113.08406,\"K\":128.09496,\"L\":113.08406,\"M\":131.04049,\"N\":114.04293,\"P\":97.05276,\"Q\":128.05858,\n",
    "              \"R\":156.10111,\"S\":87.03203,\"T\":101.04768,\"V\":99.06841,\"W\":186.07931,\"Y\":163.06333}\n",
    "\n",
    "mass = 0\n",
    "for amino in seq:\n",
    "    mass += dictionary[amino]\n",
    "\n",
    "mass = round(mass,3)\n",
    "print(mass)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 15 - Overlap Graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rosalind_0016 Rosalind_9637\n",
      "Rosalind_0016 Rosalind_8877\n",
      "Rosalind_4367 Rosalind_5215\n",
      "Rosalind_4367 Rosalind_1073\n",
      "Rosalind_4367 Rosalind_6769\n",
      "Rosalind_4367 Rosalind_2611\n",
      "Rosalind_0124 Rosalind_6807\n",
      "Rosalind_2849 Rosalind_8045\n",
      "Rosalind_2849 Rosalind_6988\n",
      "Rosalind_4918 Rosalind_6576\n",
      "Rosalind_4918 Rosalind_6269\n",
      "Rosalind_5215 Rosalind_8598\n",
      "Rosalind_5215 Rosalind_3003\n",
      "Rosalind_8045 Rosalind_2849\n",
      "Rosalind_8045 Rosalind_9327\n",
      "Rosalind_8045 Rosalind_5532\n",
      "Rosalind_6604 Rosalind_7953\n",
      "Rosalind_6604 Rosalind_2943\n",
      "Rosalind_4139 Rosalind_8598\n",
      "Rosalind_4139 Rosalind_3003\n",
      "Rosalind_3691 Rosalind_0487\n",
      "Rosalind_3691 Rosalind_2147\n",
      "Rosalind_3691 Rosalind_2307\n",
      "Rosalind_3691 Rosalind_4942\n",
      "Rosalind_4077 Rosalind_8598\n",
      "Rosalind_4077 Rosalind_3003\n",
      "Rosalind_2037 Rosalind_6807\n",
      "Rosalind_7630 Rosalind_4139\n",
      "Rosalind_7630 Rosalind_0157\n",
      "Rosalind_7630 Rosalind_3199\n",
      "Rosalind_1216 Rosalind_7953\n",
      "Rosalind_1216 Rosalind_2943\n",
      "Rosalind_7696 Rosalind_6807\n",
      "Rosalind_8870 Rosalind_8045\n",
      "Rosalind_8870 Rosalind_6988\n",
      "Rosalind_8171 Rosalind_9181\n",
      "Rosalind_8171 Rosalind_6697\n",
      "Rosalind_6687 Rosalind_5215\n",
      "Rosalind_6687 Rosalind_1073\n",
      "Rosalind_6687 Rosalind_6769\n",
      "Rosalind_6687 Rosalind_2611\n",
      "Rosalind_9676 Rosalind_5236\n",
      "Rosalind_0659 Rosalind_4524\n",
      "Rosalind_0659 Rosalind_4754\n",
      "Rosalind_9036 Rosalind_7953\n",
      "Rosalind_9036 Rosalind_2943\n",
      "Rosalind_3714 Rosalind_8045\n",
      "Rosalind_3714 Rosalind_6988\n",
      "Rosalind_8495 Rosalind_3279\n",
      "Rosalind_1943 Rosalind_0016\n",
      "Rosalind_1943 Rosalind_5101\n",
      "Rosalind_9986 Rosalind_9163\n",
      "Rosalind_9986 Rosalind_8716\n",
      "Rosalind_0290 Rosalind_7953\n",
      "Rosalind_0290 Rosalind_2943\n",
      "Rosalind_4967 Rosalind_3279\n",
      "Rosalind_0251 Rosalind_3714\n",
      "Rosalind_0251 Rosalind_0591\n",
      "Rosalind_0251 Rosalind_5795\n",
      "Rosalind_0487 Rosalind_7953\n",
      "Rosalind_0487 Rosalind_2943\n",
      "Rosalind_5101 Rosalind_3691\n",
      "Rosalind_5101 Rosalind_4096\n",
      "Rosalind_5101 Rosalind_3571\n",
      "Rosalind_9197 Rosalind_6836\n",
      "Rosalind_5268 Rosalind_6807\n",
      "Rosalind_2147 Rosalind_5236\n",
      "Rosalind_7767 Rosalind_6687\n",
      "Rosalind_5841 Rosalind_4139\n",
      "Rosalind_5841 Rosalind_0157\n",
      "Rosalind_5841 Rosalind_3199\n",
      "Rosalind_9163 Rosalind_4524\n",
      "Rosalind_9163 Rosalind_4754\n",
      "Rosalind_9637 Rosalind_0016\n",
      "Rosalind_9637 Rosalind_5101\n",
      "Rosalind_2307 Rosalind_0124\n",
      "Rosalind_2307 Rosalind_3715\n",
      "Rosalind_2307 Rosalind_8171\n",
      "Rosalind_2307 Rosalind_1253\n",
      "Rosalind_5958 Rosalind_0231\n",
      "Rosalind_5958 Rosalind_9036\n",
      "Rosalind_5958 Rosalind_9986\n",
      "Rosalind_8445 Rosalind_6604\n",
      "Rosalind_8445 Rosalind_4473\n",
      "Rosalind_7759 Rosalind_8495\n",
      "Rosalind_4473 Rosalind_8495\n",
      "Rosalind_8199 Rosalind_4077\n",
      "Rosalind_8199 Rosalind_7888\n",
      "Rosalind_9538 Rosalind_3167\n",
      "Rosalind_3571 Rosalind_0659\n",
      "Rosalind_1219 Rosalind_2037\n",
      "Rosalind_1219 Rosalind_8445\n",
      "Rosalind_1219 Rosalind_4669\n",
      "Rosalind_1253 Rosalind_8045\n",
      "Rosalind_1253 Rosalind_6988\n",
      "Rosalind_4754 Rosalind_2592\n",
      "Rosalind_9327 Rosalind_0659\n",
      "Rosalind_6697 Rosalind_5958\n",
      "Rosalind_7888 Rosalind_9163\n",
      "Rosalind_7888 Rosalind_8716\n",
      "Rosalind_3199 Rosalind_7630\n",
      "Rosalind_3199 Rosalind_9538\n",
      "Rosalind_6576 Rosalind_7953\n",
      "Rosalind_6576 Rosalind_2943\n",
      "Rosalind_6269 Rosalind_5236\n",
      "Rosalind_7361 Rosalind_8870\n",
      "Rosalind_7361 Rosalind_0290\n",
      "Rosalind_7361 Rosalind_4915\n",
      "Rosalind_7361 Rosalind_8134\n",
      "Rosalind_5532 Rosalind_4524\n",
      "Rosalind_5532 Rosalind_4754\n",
      "Rosalind_2592 Rosalind_8045\n",
      "Rosalind_2592 Rosalind_6988\n",
      "Rosalind_6807 Rosalind_9676\n",
      "Rosalind_1653 Rosalind_8870\n",
      "Rosalind_1653 Rosalind_0290\n",
      "Rosalind_1653 Rosalind_7361\n",
      "Rosalind_1653 Rosalind_4915\n",
      "Rosalind_1653 Rosalind_8134\n",
      "Rosalind_6205 Rosalind_6836\n",
      "Rosalind_4915 Rosalind_2301\n",
      "Rosalind_4915 Rosalind_1558\n",
      "Rosalind_3003 Rosalind_7630\n",
      "Rosalind_3003 Rosalind_9538\n",
      "Rosalind_2943 Rosalind_9181\n",
      "Rosalind_2943 Rosalind_6697\n",
      "Rosalind_8134 Rosalind_6604\n",
      "Rosalind_8134 Rosalind_4473\n",
      "Rosalind_1548 Rosalind_0487\n",
      "Rosalind_1548 Rosalind_2147\n",
      "Rosalind_1548 Rosalind_2307\n",
      "Rosalind_1548 Rosalind_4942\n",
      "Rosalind_2301 Rosalind_6836\n",
      "Rosalind_5236 Rosalind_8495\n",
      "Rosalind_3343 Rosalind_3279\n",
      "Rosalind_8877 Rosalind_9181\n",
      "Rosalind_8877 Rosalind_6697\n",
      "Rosalind_6769 Rosalind_4524\n",
      "Rosalind_6769 Rosalind_4754\n",
      "Rosalind_4942 Rosalind_0659\n",
      "Rosalind_1558 Rosalind_9197\n",
      "Rosalind_1607 Rosalind_7759\n",
      "Rosalind_1607 Rosalind_3343\n",
      "Rosalind_2611 Rosalind_6807\n",
      "Rosalind_8716 Rosalind_8495\n",
      "Rosalind_0591 Rosalind_4077\n",
      "Rosalind_0591 Rosalind_7888\n",
      "Rosalind_9422 Rosalind_2301\n",
      "Rosalind_9422 Rosalind_1558\n",
      "Rosalind_3167 Rosalind_3189\n"
     ]
    }
   ],
   "source": [
    "#FASTA reader:\n",
    "\n",
    "seq =''\n",
    "seqlist = [] #Create an empty list to store values - both labels and sequences \n",
    "with open(\"exercises/rosalind_grph.txt\") as file_in:\n",
    "    for line in file_in:\n",
    "        if line[0] == \">\":\n",
    "            line = line.lstrip('>')\n",
    "            line = line.rstrip('\\n')\n",
    "            seqlist.append(seq)\n",
    "            seqlist.append(line)\n",
    "            seq = \"\"\n",
    "        else:\n",
    "            line = line.rstrip('\\n')\n",
    "            seq += line\n",
    "    seqlist.append(seq) #When it runs out of lines, it'll append the last sequence\n",
    "\n",
    "seqlist.pop(0) #Remove first empty space - pop uses index directly\n",
    "\n",
    "#Code proper\n",
    "\n",
    "indexlist = [] #Create a list of indices which hold sequences in seqlist for shallow copying later\n",
    "for i in range(1, len(seqlist), 2):\n",
    "    indexlist.append(i)\n",
    "\n",
    "for i in range(0, len(indexlist)): #Sort through all sequences using the content of the list of indices\n",
    "    templist = indexlist.copy() \n",
    "    templist.pop(i) #Ensure no self-overlap, but maintain ability to search through all other sequences, both forward and backward\n",
    "    for j in range(0, len(templist)):\n",
    "        if seqlist[indexlist[i]][-3:] == seqlist[templist[j]][:3]: \n",
    "            #Matching of tail-head overlap; for every sequence, search all others both forward and backward less itself\n",
    "            print(seqlist[indexlist[i]-1], end = \" \")\n",
    "            print(seqlist[templist[j]-1])\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 16 - Finding a Protein Motif"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Webscraper works here - can use it to track financials rather than Excel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Q90304_C166_CARAU\n",
      "92 171 350 441 465 \n",
      "P08318_P100_HCMVA\n",
      "833 \n",
      "Q5FTZ8\n",
      "49 62 \n",
      "P01047_KNL2_BOVIN\n",
      "47 87 168 169 197 204 280 \n",
      "Q2YCH6\n",
      "249 \n",
      "P02974_FMM1_NEIGO\n",
      "67 68 121 \n",
      "O82484\n",
      "104 108 546 742 765 \n",
      "Q9V730\n",
      "71 327 476 \n",
      "Q7S432\n",
      "173 \n",
      "P43541\n",
      "129 \n"
     ]
    }
   ],
   "source": [
    "from bs4 import BeautifulSoup\n",
    "import requests\n",
    "\n",
    "#Open file as usual\n",
    "linelist = []\n",
    "with open(\"exercises/rosalind_mprt (1).txt\") as file_in:\n",
    "    for line in file_in:\n",
    "        line = line.rstrip(\"\\n\")\n",
    "        linelist.append(line)\n",
    "        \n",
    "#First, we need to turn the IDs into url form:\n",
    "urllist = []\n",
    "for i in range(0, len(linelist)):\n",
    "    skeletonhead = \"http://www.uniprot.org/uniprot/\"\n",
    "    skeletontail = \".fasta\"\n",
    "    url = skeletonhead + linelist[i] + skeletontail\n",
    "    urllist.append(url)\n",
    "        \n",
    "#Then, we'll scrape the relevant fasta pages into a txt file\n",
    "f = open(\"exercises/fastastorage.txt\", \"w\") #we overwrite any existing content here\n",
    "for i in range(0, len(urllist)):\n",
    "    html_content = requests.get(urllist[i]).text\n",
    "    soup = BeautifulSoup(html_content, \"lxml\")\n",
    "    f.write(soup.body.text)\n",
    "f.close()\n",
    "\n",
    "#open with our fasta reader (honestly there wasn't a need to create a new file)\n",
    "seq =''\n",
    "seqlist = [] #Create an empty list to store values - both labels and sequences \n",
    "with open(\"exercises/fastastorage.txt\") as f:\n",
    "    for line in f:\n",
    "        if line[0] == \">\":\n",
    "            line = line.lstrip('>')\n",
    "            line = line.rstrip('\\n')\n",
    "            seqlist.append(seq)\n",
    "            seqlist.append(line)\n",
    "            seq = \"\"\n",
    "        else:\n",
    "            line = line.rstrip('\\n')\n",
    "            seq += line\n",
    "    seqlist.append(seq) #When it runs out of lines, it'll append the last sequence\n",
    "seqlist.pop(0) #Remove first empty space\n",
    "\n",
    "#Then, we'll scan for all possible permutations of the N-glycosylation motif and store these locations in a list, along with their title\n",
    "motifloc = []\n",
    "for i in range(1, len(seqlist), 2):\n",
    "    motifloc.append(seqlist[i-1])\n",
    "    locholder = \"\"\n",
    "    for j in range(0, len(seqlist[i])-3):\n",
    "        if seqlist[i][j] == \"N\":\n",
    "            if seqlist[i][j+1] != \"P\":\n",
    "                if seqlist[i][j+2] == \"S\" or seqlist[i][j+2] == \"T\":\n",
    "                    if seqlist[i][j+3] != \"P\":\n",
    "                        locholder += str(j+1) + \" \" #Remember that indices begin with 0 but amino acid positions start with 1\n",
    "    motifloc.append(locholder)\n",
    "\n",
    "#We then create a final list consisting of the IDs from the original file and the scraped motif location data\n",
    "finallist = [\"\"]*len(motifloc) #This list is of the same length as motifloc\n",
    "for i in range(0, len(finallist)):\n",
    "    if i%2 == 0:\n",
    "        finallist[i] = linelist[int(i/2)] #Every even index, 0 included, is a Uniprot ID, and we know linelist is half the length of finallist\n",
    "    else:\n",
    "        finallist[i] = motifloc[i]\n",
    "        if motifloc[i] != \"\": #If it isn't devoid of the motif, I want to print both it and the ID\n",
    "            print(finallist[i-1])\n",
    "            print(motifloc[i])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 17 - Consensus and Profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TAAGTTCGATACATTATACACCAGCACAGCAAACCCTTACCGTCGCAAAGGTAACAGCTGGTTTCAGATCGACGCCTTCCTGGAGCCCACCCGCAACAATCCTAACGATAAAGTGCATGGCAAGAAGACTCCGATACGACCAAGATTAAATATGGAAGGTAGCATGGAACCGCGGGCGTTAACCACAGATCCCCCCTCATTCCAGACATATATTCAAACTTAACTGCACGCGACTGGCAACGAGGCTTTATCACTTACGCCGAATCCTTCGTACGGTGGTCTAATCTATGACTGGAGAGTCATAAAAACAAAGCACGACGCACGAGATCACAGACAGGAAGACAAACAGTGAGATAAATCATAACTTACCACCCCAAGTTGCGCATTACAGGAGCCCAAATCATGTCCGGTAGCTCAGTCACTCTAACATGGGGTGCGTCACCAGTGACTGTAGGCCAGTATTAACAGATCTCCATCGATTACAAAAACGACGCGACATCGGCGCTAATCGATAGACCACGATGGCTTTAATTCAAACCTACAAAAAAGAATTCCCGACAGGTCACGGGCGCAGGTATGTCCACGCAAAAGTCTCACGAACCCTTGTGTGGTATCCACCTTATATATTGAGTGCAGCCGTACGGTTCCCAAGGCAATAATGCTTGTCGGCGTACCGGCTATGCAACATTCAAAAACAGCGAGACATCAAATTCAGGTGGAGGTGGCAAGCAGGCGTTCCCAACAGTCATGCGACCTATTAAAGCGGAACCAACGGAGTACAGCATGAACATACCAATACGCACCATAGAAAACCGCTAACGGGCATAAGCCTGTTGCTTATGGTCCGAGACAAACAGGAGCGTACCTATGCAGACTAACGGATTCACCTGCAGACCTTACACTAGCACACCATTACTACGTAAAAAAAATACACCGGACGGTAACGACGGTA\n",
      "A: 3 4 3 3 1 3 0 3 3 2 5 2 3 2 3 3 2 4 2 3 0 1 4 2 2 5 1 5 2 1 3 3 5 3 3 3 1 1 5 3 1 3 1 2 2 1 5 4 4 2 2 2 4 4 3 3 1 2 3 2 2 1 3 1 1 3 2 6 1 2 3 4 2 2 2 3 1 0 2 3 1 2 3 3 1 0 1 2 5 2 3 2 3 3 4 4 2 3 5 1 1 3 3 3 5 2 3 3 3 3 6 5 2 2 2 2 3 3 2 2 1 7 3 2 4 3 2 5 2 2 2 0 2 4 1 4 2 1 4 2 2 4 4 4 5 1 1 5 4 3 0 3 1 1 2 5 5 2 2 2 4 2 2 4 2 2 2 3 3 3 2 2 1 2 1 2 1 3 2 3 4 4 1 2 5 3 3 2 4 1 2 2 2 2 2 1 1 2 4 2 3 3 2 4 1 5 4 3 2 5 2 4 2 1 0 4 3 4 2 1 2 3 4 2 1 2 3 5 2 2 2 2 4 1 3 3 2 2 4 5 1 0 4 1 1 2 2 2 2 3 1 0 5 2 0 2 4 1 2 2 1 1 3 3 3 2 3 2 1 2 2 3 5 1 1 3 2 2 0 3 2 2 6 4 1 3 2 5 1 1 5 3 2 3 0 4 0 3 3 2 0 5 0 3 3 3 6 4 1 4 4 3 2 0 3 1 1 5 2 1 2 4 2 1 4 3 4 0 2 3 0 4 3 3 1 6 2 1 3 3 1 3 2 3 7 4 2 4 1 2 2 5 2 6 2 4 3 3 2 2 3 3 3 5 1 3 2 4 3 2 5 3 2 1 1 4 4 1 1 2 0 0 3 0 3 2 2 5 1 3 0 2 3 0 1 1 2 3 4 4 2 0 4 3 3 1 2 3 2 1 2 4 2 1 0 2 4 1 2 1 4 1 3 0 3 5 4 2 4 3 3 1 2 2 3 2 0 2 2 2 3 1 1 3 1 0 1 5 1 1 3 3 3 2 3 2 2 5 3 3 3 1 1 3 3 3 4 1 4 4 2 2 3 1 3 2 2 3 3 0 2 3 2 6 6 5 4 4 2 1 3 3 1 1 2 4 2 3 1 3 2 2 2 1 1 2 4 5 0 1 3 5 3 4 0 6 2 1 3 2 2 3 2 2 0 3 1 3 3 5 4 2 2 1 4 4 3 3 3 1 4 2 3 4 3 4 4 3 3 3 5 1 2 2 2 1 2 4 1 5 3 1 1 3 3 0 0 1 0 2 2 2 4 1 3 2 3 2 1 1 1 1 5 2 1 2 4 3 4 3 1 2 1 1 3 3 2 3 4 3 1 2 2 1 2 0 1 2 1 3 3 1 5 3 2 1 5 2 2 2 1 3 2 4 3 4 3 2 1 3 2 2 2 2 3 2 1 1 2 2 5 1 2 2 1 1 2 2 1 4 4 2 2 2 5 3 2 3 3 2 3 1 2 3 2 2 2 2 1 0 1 1 3 1 1 0 1 1 2 4 2 1 2 4 4 0 3 2 1 2 4 4 4 5 4 2 4 0 2 3 4 1 3 2 4 3 2 3 4 3 1 1 1 4 2 2 1 3 2 3 1 1 1 1 2 2 3 3 2 2 4 2 2 2 0 1 1 3 1 2 3 3 2 4 0 1 2 5 2 1 2 1 3 2 2 2 4 2 4 5 3 3 3 2 0 3 4 4 1 2 3 3 2 2 2 3 0 2 4 0 4 3 1 4 3 2 4 3 3 3 3 3 2 0 7 4 1 5 2 1 3 3 1 2 4 3 4 1 3 5 4 5 3 2 2 0 3 4 5 3 3 2 1 0 5 1 3 3 1 2 3 1 3 0 1 2 1 1 1 4 1 2 2 1 2 2 2 4 3 4 2 3 3 3 1 3 3 1 5 2 1 2 0 3 2 3 1 3 2 2 2 3 1 3 2 3 3 4 3 2 3 3 2 3 2 4 3 2 1 2 2 4 2 3 2 2 1 1 3 3 5 2 0 7 3 1 3 2 5 2 1 4 0 2 5 3 1 4 2 2 3 4 3 5 5 3 4 4 4 3 4 1 3 1 2 2 1 5 2 2 2 1 4 4 1 2 4 2 3 1 1 3\n",
      "C: 2 3 2 3 2 1 4 3 3 3 3 3 2 1 0 3 2 4 5 2 5 4 2 3 3 1 4 2 2 5 2 2 3 4 4 4 1 3 1 4 4 1 4 3 2 4 1 3 3 2 3 1 2 1 4 3 2 3 0 2 3 2 1 2 6 3 2 1 3 3 1 1 4 2 3 4 2 3 5 5 0 2 2 2 3 4 4 4 1 3 4 3 2 4 4 1 3 3 2 3 5 4 3 3 1 3 1 2 0 3 2 2 2 1 2 4 3 0 2 1 5 2 3 1 1 2 1 1 5 2 4 6 2 1 2 3 4 0 2 4 4 3 1 0 1 1 2 2 2 3 2 3 2 2 2 2 0 1 3 1 3 2 5 2 2 1 2 3 3 6 3 3 5 2 1 2 4 0 3 1 2 1 4 4 2 4 1 0 2 2 5 4 4 4 4 4 2 3 0 2 1 5 6 2 2 2 5 1 2 2 3 3 3 2 4 2 3 3 3 3 1 3 2 5 2 2 5 2 4 0 3 2 4 3 1 1 1 3 3 1 4 3 3 2 1 3 2 1 0 3 3 5 2 5 1 1 1 3 2 4 5 3 1 3 2 4 6 1 2 4 2 1 1 4 1 0 2 1 1 0 3 3 0 3 3 4 1 1 2 2 1 5 3 1 3 3 2 3 1 3 4 5 3 1 3 2 4 2 3 3 2 3 1 4 2 4 3 2 3 2 4 2 3 2 2 2 2 2 4 2 4 3 2 3 3 2 2 2 3 1 3 2 3 3 0 4 4 4 3 2 2 2 1 2 3 2 3 3 2 3 2 0 2 1 3 1 0 0 4 4 4 5 4 5 5 2 3 1 2 1 3 5 1 7 2 3 2 0 4 2 1 2 3 1 5 3 4 3 3 2 2 4 1 1 0 2 4 5 3 2 3 3 3 4 2 3 3 2 2 4 2 4 0 4 0 1 3 4 0 2 2 2 0 2 2 2 4 2 1 3 2 4 3 2 1 3 1 1 5 1 1 1 3 1 2 3 4 1 2 1 2 2 2 3 2 4 0 3 3 1 4 2 4 7 2 3 4 1 2 2 2 2 3 1 1 3 3 1 6 2 3 4 2 5 2 1 3 3 2 4 2 2 3 1 4 1 3 3 2 4 1 1 0 2 4 0 5 4 2 4 3 3 2 2 2 4 3 0 0 3 1 1 3 4 2 2 2 4 5 1 2 5 2 3 3 2 1 2 3 1 3 1 1 3 5 4 2 1 5 1 0 1 3 4 2 4 2 2 0 3 2 3 3 3 2 2 3 2 1 3 5 3 3 4 0 3 1 3 2 3 3 2 5 3 4 2 6 2 2 2 3 4 4 3 2 1 3 2 2 1 1 2 1 1 3 4 2 3 4 0 2 2 3 4 0 1 1 3 1 3 2 2 3 3 2 2 4 3 1 1 2 3 2 0 3 1 4 5 4 3 3 3 1 3 1 3 3 3 1 2 2 5 2 0 1 1 4 2 2 5 3 2 2 4 3 1 1 4 3 2 1 3 5 3 4 5 3 3 2 3 0 2 3 1 1 4 1 2 4 2 1 3 2 4 1 1 4 2 1 2 3 3 6 3 1 2 3 1 1 1 0 3 2 0 1 4 3 1 2 5 0 2 1 3 2 1 1 5 5 3 2 2 4 1 4 2 4 0 3 2 3 3 3 4 5 1 2 2 1 2 3 2 1 5 3 1 1 2 5 3 2 1 4 2 2 2 2 1 2 4 3 2 4 1 2 1 2 3 5 3 2 3 3 8 2 2 2 0 4 1 4 1 4 4 2 3 1 1 3 1 1 2 4 4 2 5 1 1 3 4 1 2 1 5 3 2 3 3 2 4 4 1 3 3 1 2 3 3 3 3 1 1 2 4 6 4 2 2 1 3 3 2 1 3 5 2 2 3 5 2 3 2 3 3 4 4 2 2 2 1 4 2 2 3 3 1 3 2 4 1 1 3 2 0 3 4 6 4 3 2 4 1 3 3 4 4 3 2 1 5 1 5 2 1 1 5 3 4 1 4 3 3 4 2 0 4 3 3 4 3 2 0 2 3 1 2 2 2 3 0 3 4 3 3 4 2 3 2 4 2 2 3 3 2 5 2 3 4 2 1 3 2\n",
      "G: 0 2 2 4 3 2 3 4 3 1 1 2 3 1 3 3 2 1 2 2 2 2 3 4 2 4 3 3 4 3 2 3 1 1 0 2 2 2 2 0 3 5 0 3 3 3 1 0 2 3 4 3 1 2 1 3 4 2 2 5 5 3 2 2 1 3 4 1 2 3 4 3 3 3 2 2 2 2 2 1 2 3 5 3 4 3 4 2 2 2 3 3 4 1 1 3 2 2 3 2 0 0 0 2 2 3 4 3 3 2 1 2 3 2 4 2 3 2 3 4 1 1 3 4 3 3 4 2 2 2 2 4 3 3 3 0 2 6 3 2 1 2 2 5 1 2 3 2 2 1 2 1 2 4 3 3 4 4 4 3 1 4 1 2 1 4 5 3 2 0 3 4 2 4 4 3 4 4 1 2 2 1 1 2 3 3 3 6 0 3 1 1 2 2 1 4 1 2 2 1 0 1 2 2 4 3 1 3 2 0 1 2 1 3 3 3 3 2 2 2 3 3 1 1 3 4 1 2 2 4 3 5 1 3 1 4 5 2 1 3 2 4 2 6 6 3 1 3 1 3 2 5 3 2 2 2 2 3 6 3 2 4 3 1 1 2 1 1 2 3 4 2 2 2 4 4 2 4 8 2 3 1 2 1 2 2 2 3 3 4 1 1 0 4 6 1 5 1 4 0 3 0 3 3 1 2 0 2 3 2 3 2 5 3 2 3 4 2 2 4 2 2 2 6 2 4 1 3 2 3 4 3 4 2 3 0 4 6 2 3 4 2 3 2 2 1 1 1 4 2 5 2 5 1 1 3 2 3 2 3 3 3 2 2 3 2 2 3 0 1 1 2 4 4 2 2 1 4 3 3 4 2 4 1 2 1 2 2 2 3 7 3 3 5 2 3 2 1 1 3 2 4 3 1 5 3 4 2 5 4 1 0 4 1 2 3 1 5 1 3 2 1 3 3 3 2 0 1 3 1 5 6 4 3 1 4 4 4 2 2 3 3 3 3 4 3 5 2 1 3 4 2 3 4 4 3 1 2 5 2 2 2 2 2 2 2 3 4 3 0 3 2 3 1 3 1 3 4 2 3 1 3 2 3 2 1 3 4 1 4 1 1 4 1 4 3 2 1 3 0 3 4 2 4 3 3 0 1 2 2 4 2 3 2 5 0 3 2 3 3 4 2 2 4 4 1 1 3 1 0 1 1 1 1 1 3 2 2 1 2 1 1 3 2 2 1 3 3 4 3 1 3 3 3 1 4 3 2 2 3 4 7 2 3 2 4 5 5 6 2 3 2 3 4 4 2 3 1 5 2 3 3 0 1 5 3 2 1 1 1 4 1 1 2 0 2 2 4 3 2 3 2 2 0 2 5 1 5 3 4 5 3 3 2 2 4 1 2 4 2 2 2 1 0 2 2 2 1 5 2 5 2 4 3 2 5 1 3 4 3 2 3 4 5 2 2 3 2 4 2 1 4 5 2 4 3 1 1 3 2 4 2 2 2 4 3 3 3 4 2 4 2 3 4 3 7 6 2 1 2 3 4 1 1 1 0 3 1 3 2 3 2 1 1 2 1 3 6 2 4 3 4 2 1 4 2 2 3 1 3 0 2 2 3 4 4 2 6 6 3 5 5 2 7 5 1 1 3 3 1 2 5 5 3 4 3 2 0 3 3 3 3 3 3 5 3 3 3 1 5 2 4 1 1 2 2 0 2 0 3 2 2 4 1 4 4 2 2 2 3 2 3 4 3 3 2 6 3 2 3 2 4 2 3 1 6 3 2 1 2 1 2 3 1 1 2 3 0 4 6 2 3 4 3 2 0 3 5 3 1 2 0 1 2 4 2 1 4 0 0 4 4 6 4 1 1 1 3 5 1 1 3 4 3 2 3 3 1 2 3 3 4 4 0 0 3 4 4 5 1 2 3 3 2 1 3 4 4 0 4 3 3 3 3 0 1 3 3 2 4 2 3 4 2 3 2 2 3 1 4 4 2 2 3 2 1 1 3 0 3 1 3 4 1 4 2 2 3 3 1 2 3 2 0 6 1 3 1 1 2 3 2 1 2 2 2 1 2 3 4 1 2 3 1 3 3 0 3 2 3 0 2 3 3 0 4 6 1 2 5 3 2 1 1 0 3 1 3 4 4 2 3\n",
      "T: 5 1 3 0 4 4 3 0 1 4 1 3 2 6 4 1 4 1 1 3 3 3 1 1 3 0 2 0 2 1 3 2 1 2 3 1 6 4 2 3 2 1 5 2 3 2 3 3 1 3 1 4 3 3 2 1 3 3 5 1 0 4 4 5 2 1 2 2 4 2 2 2 1 3 3 1 5 5 1 1 7 3 0 2 2 3 1 2 2 3 0 2 1 2 1 2 3 2 0 4 4 3 4 2 2 2 2 2 4 2 1 1 3 5 2 2 1 5 3 3 3 0 1 3 2 2 3 2 1 4 2 0 3 2 4 3 2 3 1 2 3 1 3 1 3 6 4 1 2 3 6 3 5 3 3 0 1 3 1 4 2 2 2 2 5 3 1 1 2 1 2 1 2 2 4 3 1 3 4 4 2 4 4 2 0 0 3 2 4 4 2 3 2 2 3 1 6 3 4 5 6 1 0 2 3 0 0 3 4 3 4 1 4 4 3 1 1 1 3 4 4 1 3 2 4 2 1 1 2 4 2 1 1 3 5 2 2 3 2 1 3 3 1 1 2 2 5 4 7 1 4 0 0 1 7 5 3 3 0 1 2 2 3 3 4 2 0 6 5 1 2 4 2 3 4 3 4 3 1 5 2 4 2 2 4 1 5 1 4 3 3 1 5 2 1 2 3 3 2 5 3 0 4 3 3 3 0 2 3 1 1 2 2 3 3 2 2 1 3 3 2 2 3 1 2 1 3 5 2 2 2 0 1 2 3 2 2 1 2 3 2 3 2 2 1 1 3 1 2 4 1 1 2 1 4 1 2 1 4 2 2 4 3 2 3 4 6 3 3 3 0 0 0 0 2 2 2 4 4 4 3 3 2 2 3 4 4 3 3 2 2 3 1 4 2 3 2 3 2 1 4 2 2 5 2 4 0 0 0 3 4 3 1 4 6 2 2 2 5 2 2 4 4 3 4 2 3 3 3 4 0 1 4 3 4 2 2 2 5 3 2 2 3 2 4 4 3 2 3 5 2 4 1 3 1 2 3 2 0 4 3 5 5 2 3 1 3 2 0 5 1 4 0 1 2 4 1 2 3 5 5 2 3 0 1 1 0 1 1 3 3 2 3 3 2 2 3 3 4 3 3 2 3 4 2 4 3 1 6 3 2 2 4 2 1 4 0 3 2 1 1 2 4 2 4 2 5 4 6 2 4 6 4 4 3 1 3 1 1 6 3 2 2 1 2 3 2 2 0 3 1 5 4 2 2 1 3 3 2 1 3 1 4 0 3 2 3 2 4 3 3 3 0 2 1 4 1 5 3 4 1 3 2 3 4 2 3 3 3 3 2 5 3 4 3 3 0 1 1 3 3 2 2 6 4 4 5 1 4 2 1 4 1 4 3 1 2 3 0 6 5 3 4 2 5 3 4 4 3 2 1 4 1 2 3 1 4 3 3 4 1 3 2 3 4 6 1 1 1 1 2 1 2 3 0 1 4 3 3 4 1 2 4 5 3 4 1 3 3 3 2 5 2 1 3 2 2 3 4 2 4 2 2 2 1 5 1 4 4 3 3 2 2 3 3 3 2 2 2 1 2 2 3 3 1 4 2 2 4 2 6 4 1 0 3 2 4 0 1 3 4 1 5 2 2 3 3 3 3 2 4 1 2 2 4 5 6 2 1 2 2 2 1 2 1 4 1 2 4 2 3 2 3 3 1 5 4 4 5 0 2 3 2 2 3 2 3 2 2 2 3 3 0 3 3 3 2 4 2 3 1 1 3 2 4 1 1 2 1 2 4 2 2 1 0 2 4 5 0 2 1 3 1 1 2 4 2 3 1 3 3 3 2 2 2 3 5 1 2 3 2 2 2 1 1 6 3 1 2 3 2 5 0 4 6 3 3 5 4 0 5 3 2 5 2 1 2 0 1 2 3 2 3 2 3 2 1 2 0 2 3 3 4 1 4 2 4 2 4 3 2 2 3 2 2 4 2 1 2 3 2 2 4 4 3 1 0 1 6 3 3 2 1 3 0 2 4 4 3 1 2 0 6 2 0 3 1 3 3 2 3 1 5 4 3 1 5 1 1 1 4 4 2 1 1 2 4 1 1 4 3 3 1 3 4 2 0 2 2 1 3 4 2 3 4 3 2 1 1 4 4 2\n"
     ]
    }
   ],
   "source": [
    "seq =''\n",
    "seqlist = [] #Create an empty list to store values - both labels and sequences \n",
    "with open(\"exercises/rosalind_cons.txt\") as file_in:\n",
    "    for line in file_in:\n",
    "        if line[0] == \">\":\n",
    "            line = line.lstrip('>')\n",
    "            line = line.rstrip('\\n')\n",
    "            seqlist.append(seq)\n",
    "            seqlist.append(line)\n",
    "            seq = \"\"\n",
    "        else:\n",
    "            line = line.rstrip('\\n')\n",
    "            seq += line\n",
    "    seqlist.append(seq) #When it runs out of lines, it'll append the last sequence\n",
    "seqlist.pop(0) #Remove first empty space\n",
    "\n",
    "n = len(seqlist[1])\n",
    "\n",
    "#You cannot simply multiply [[0]*n]*4 as they will all be dependent copies of each other (Python holds references)\n",
    "\n",
    "A = [0] * n \n",
    "C = [0] * n  \n",
    "G = [0] * n \n",
    "T = [0] * n \n",
    "\n",
    "matrix = [A,C,G,T]\n",
    "\n",
    "#For whatever particular nucleotide, we add 1 to the relevant position in the matrix\n",
    "for i in range(1, len(seqlist), 2): \n",
    "    for j in range(0, len(seqlist[1])):\n",
    "        if seqlist[i][j] == \"A\":\n",
    "            matrix[0][j] += 1\n",
    "        elif seqlist[i][j] == \"C\":\n",
    "            matrix[1][j] += 1\n",
    "        elif seqlist[i][j] == \"G\":\n",
    "            matrix[2][j] += 1\n",
    "        elif seqlist[i][j] == \"T\":\n",
    "            matrix[3][j] += 1\n",
    "            \n",
    "cons = \"\" #Initialise consensus string and dictionary\n",
    "dictionary = {0:\"A\", 1: \"C\", 2:\"G\", 3:\"T\"}\n",
    "\n",
    "#Now, we want to vertically traverse the matrix and return the index of the maximum value, as this will inform our consensus sequence\n",
    "for i in range(0, n):\n",
    "    vertcompare = []\n",
    "    for j in range(0,4):\n",
    "        vertcompare.append(matrix[j][i]) #We append the value at the ith position of all the lists A/C/G/T in sequence\n",
    "    consnuc = vertcompare.index(max(vertcompare)) #Index of maximum value will tell us what the consensus nucleotide at that position is\n",
    "    cons += dictionary[consnuc]\n",
    "\n",
    "print(cons)\n",
    "\n",
    "print(\"A: \", end = \"\")\n",
    "print(*matrix[0], sep = \" \")\n",
    "print(\"C: \", end = \"\")\n",
    "print(*matrix[1], sep = \" \")\n",
    "print(\"G: \", end = \"\")\n",
    "print(*matrix[2], sep = \" \")\n",
    "print(\"T: \", end = \"\")\n",
    "print(*matrix[3], sep = \" \")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 18 - Open Reading Frames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MLYSFPPHLPTSCTT\n",
      "MWVSFKTSKVSCGDR\n",
      "MPRGHLGTGIRPNVADR\n",
      "MVFEAVGTGRE\n",
      "MLLAWSTETLDRNDSSLGFALFQRLL\n",
      "MT\n",
      "MSSLSDCL\n",
      "METALSTPQSNSPKVLMLTRDCMKRNDILAFH\n",
      "MASSSRLRAPFQVSYYGPTSLQVQIGNLASLYDSPVRKRQG\n",
      "MLGHNSSPETALLVVTTKPFKSCTRWANVAGTNTAFGHVNDHHTKLYSF\n",
      "MSPCC\n",
      "MRLLA\n",
      "MTSGH\n",
      "MMDSLLTAEGLISLIRAQYLPPQIP\n",
      "MLTRDCMKRNDILAFH\n",
      "MWRERIQHLATLTITTRNFTRFERYPH\n",
      "MTRLWYLRR\n",
      "MAQHLFKSRLETWLPSMIHRSVKGRDKPYVLSPWSEVNRRNLGKGSVAAYRQSERDDIGAFQYALLLASHDASSGVI\n",
      "MKRNDILAFH\n",
      "MIHRSVKGRDKPYVLSPWSEVNRRNLGKGSVAAYRQSERDDIGAFQYALLLASHDASSGVI\n",
      "MAT\n",
      "MDSLLTAEGLISLIRAQYLPPQIP\n",
      "MMRLLA\n",
      "MPLMPRGHLGTGIRPNVADR\n",
      "MREIRPSAVNRESIIFWRLNS\n",
      "MIRRTV\n",
      "MAG\n"
     ]
    }
   ],
   "source": [
    "seq =''\n",
    "seqlist = [] #Create an empty list to store values - both labels and sequences \n",
    "with open(\"exercises/rosalind_orf (1).txt\") as file_in:\n",
    "    for line in file_in:\n",
    "        if line[0] == \">\":\n",
    "            line = line.lstrip('>')\n",
    "            line = line.rstrip('\\n')\n",
    "            seqlist.append(seq)\n",
    "            seqlist.append(line)\n",
    "            seq = \"\"\n",
    "        else:\n",
    "            line = line.rstrip('\\n')\n",
    "            seq += line\n",
    "    seqlist.append(seq) #When it runs out of lines, it'll append the last sequence\n",
    "seqlist.pop(0) #Remove first empty space\n",
    "\n",
    "#First, we translate to RNA\n",
    "rnaseq = \"\"\n",
    "for nuc in seqlist[1]:\n",
    "    if nuc == 'T':\n",
    "        rnaseq += 'U'\n",
    "    else:\n",
    "        rnaseq += nuc\n",
    "\n",
    "#And obtain reverse complement\n",
    "dictionarynuc = {'A':'U', 'U':'A', 'C':'G', 'G':'C'}\n",
    "crnaseq = \"\"\n",
    "for nuc in rnaseq:\n",
    "    crnaseq += dictionarynuc[nuc]\n",
    "crnaseq = crnaseq[::-1] #Reverse complement in RNA\n",
    "\n",
    "dictionary = {\"UUU\" : \"F\",\"CUU\" : \"L\",\"AUU\" : \"I\",\"GUU\" : \"V\",\"UUC\" : \"F\",\"CUC\" : \"L\",\"AUC\" : \"I\",\"GUC\" : \"V\",\"UUA\" : \"L\",\"CUA\" : \"L\",\n",
    "              \"AUA\" : \"I\",\"GUA\" : \"V\",\"UUG\" : \"L\",\"CUG\" : \"L\",\"AUG\" : \"M\",\"GUG\" : \"V\",\"UCU\" : \"S\",\"CCU\" : \"P\",\"ACU\" : \"T\",\"GCU\" : \"A\",\n",
    "              \"UCC\" : \"S\",\"CCC\" : \"P\",\"ACC\" : \"T\",\"GCC\" : \"A\",\"UCA\" : \"S\",\"CCA\" : \"P\",\"ACA\" : \"T\",\"GCA\" : \"A\",\"UCG\" : \"S\",\"CCG\" : \"P\",\n",
    "              \"ACG\" : \"T\",\"GCG\" : \"A\",\"UAU\" : \"Y\",\"CAU\" : \"H\",\"AAU\" : \"N\",\"GAU\" : \"D\",\"UAC\" : \"Y\",\"CAC\" : \"H\",\"AAC\" : \"N\",\"GAC\" : \"D\",\n",
    "              \"UAA\" : \" \",\"CAA\" : \"Q\",\"AAA\" : \"K\", \"GAA\" : \"E\",\"UAG\" : \" \",\"CAG\" : \"Q\",\"AAG\" : \"K\",\"GAG\" : \"E\",\"UGU\" : \"C\",\"CGU\" : \"R\",\n",
    "              \"AGU\" : \"S\",\"GGU\" : \"G\",\"UGC\" : \"C\",\"CGC\" : \"R\",\"AGC\" : \"S\",\"GGC\" : \"G\",\"UGA\" : \" \",\"CGA\" : \"R\",\"AGA\" : \"R\",\"GGA\" : \"G\",\n",
    "              \"UGG\" : \"W\",\"CGG\" : \"R\",\"AGG\" : \"R\",\"GGG\" : \"G\"}\n",
    "\n",
    "sequences = [rnaseq, crnaseq]\n",
    "\n",
    "trans = [] #List to store protein sequences\n",
    "for h in range(0, 2):\n",
    "    for i in range(0, len(sequences[h])):\n",
    "        if sequences[h][i:i+3] == \"AUG\": #When a start codon is identified, begin search\n",
    "            translated = \"\"\n",
    "            for j in range(i, len(sequences[h]), 3): #This search will take place in blocks of 3\n",
    "                if j == len(sequences[h])-1 or j == len(sequences[h])-2: #If we encounter the end before a stop codon, discard it\n",
    "                    translated = \"\"\n",
    "                    break\n",
    "                if dictionary[sequences[h][j:j+3]] == \" \": #When we encounter a stop codon, save it\n",
    "                    trans.append(translated)\n",
    "                    translated = \"\"\n",
    "                    break\n",
    "                translated += dictionary[sequences[h][j:j+3]] #if it's not an end or a stop, continue translating\n",
    "            translated = \"\" #If we end exactly on an amino acid, it's not useful - discard it\n",
    "            \n",
    "uniquetrans = list(set(trans)) #Obtain a list of unique values; sets are not indexable\n",
    "\n",
    "for i in range(0, len(uniquetrans)):\n",
    "    print(uniquetrans[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 19 - RNA Splicing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MQGRNAGTSTIFLSTVSRISYMVFHRSRYAYRYDFRVQVHLRLLWERRESELIFRSEMPLSVIHTNPPEPSPKVLWATLLPQADASLRYYLILNRYQFQFIFVLPRLYRSGCPHRATVGDVSVARAVRGLAKVERLAYNGDNMSYNNAYETVCTLGPREIGHAVHMNLQKSIHEREPCITISCFLEPFSNGVGPGPNRPNPC \n"
     ]
    }
   ],
   "source": [
    "seq =''\n",
    "seqlist = [] #Create an empty list to store values - both labels and sequences \n",
    "with open(\"exercises/rosalind_splc.txt\") as file_in:\n",
    "    for line in file_in:\n",
    "        if line[0] == \">\":\n",
    "            line = line.lstrip('>')\n",
    "            line = line.rstrip('\\n')\n",
    "            seqlist.append(seq)\n",
    "            seqlist.append(line)\n",
    "            seq = \"\"\n",
    "        else:\n",
    "            line = line.rstrip('\\n')\n",
    "            seq += line\n",
    "    seqlist.append(seq) #When it runs out of lines, it'll append the last sequence\n",
    "seqlist.pop(0) #Remove first empty space\n",
    "\n",
    "dnaseq = seqlist[1]\n",
    "\n",
    "introns = []\n",
    "for i in range(3, len(seqlist), 2):\n",
    "    introns.append(seqlist[i])\n",
    "\n",
    "    \n",
    "for i in range(0, len(introns)):\n",
    "    n = len(introns[i])\n",
    "    for j in range(0, len(dnaseq)):\n",
    "        if dnaseq[j:j+n] == introns[i]:\n",
    "            dnaseq = dnaseq.replace(dnaseq[j:j+n], \"\") \n",
    "            \n",
    "dictionary = {\"UUU\" : \"F\",\"CUU\" : \"L\",\"AUU\" : \"I\",\"GUU\" : \"V\",\"UUC\" : \"F\",\"CUC\" : \"L\",\"AUC\" : \"I\",\"GUC\" : \"V\",\"UUA\" : \"L\",\"CUA\" : \"L\",\n",
    "              \"AUA\" : \"I\",\"GUA\" : \"V\",\"UUG\" : \"L\",\"CUG\" : \"L\",\"AUG\" : \"M\",\"GUG\" : \"V\",\"UCU\" : \"S\",\"CCU\" : \"P\",\"ACU\" : \"T\",\"GCU\" : \"A\",\n",
    "              \"UCC\" : \"S\",\"CCC\" : \"P\",\"ACC\" : \"T\",\"GCC\" : \"A\",\"UCA\" : \"S\",\"CCA\" : \"P\",\"ACA\" : \"T\",\"GCA\" : \"A\",\"UCG\" : \"S\",\"CCG\" : \"P\",\n",
    "              \"ACG\" : \"T\",\"GCG\" : \"A\",\"UAU\" : \"Y\",\"CAU\" : \"H\",\"AAU\" : \"N\",\"GAU\" : \"D\",\"UAC\" : \"Y\",\"CAC\" : \"H\",\"AAC\" : \"N\",\"GAC\" : \"D\",\n",
    "              \"UAA\" : \" \",\"CAA\" : \"Q\",\"AAA\" : \"K\", \"GAA\" : \"E\",\"UAG\" : \" \",\"CAG\" : \"Q\",\"AAG\" : \"K\",\"GAG\" : \"E\",\"UGU\" : \"C\",\"CGU\" : \"R\",\n",
    "              \"AGU\" : \"S\",\"GGU\" : \"G\",\"UGC\" : \"C\",\"CGC\" : \"R\",\"AGC\" : \"S\",\"GGC\" : \"G\",\"UGA\" : \" \",\"CGA\" : \"R\",\"AGA\" : \"R\",\"GGA\" : \"G\",\n",
    "              \"UGG\" : \"W\",\"CGG\" : \"R\",\"AGG\" : \"R\",\"GGG\" : \"G\"}\n",
    "\n",
    "rnaseq = dnaseq.replace(\"T\", \"U\")\n",
    "\n",
    "prot = \"\"\n",
    "for i in range(0, len(rnaseq),3):\n",
    "    prot += dictionary[rnaseq[i:i+3]]\n",
    "\n",
    "print(prot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 20 - Finding a Shared Motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCATGGCTCAGAATCCGCGAGGTGTGGCACCTCACCTTCAACGAATCAGGCTGTCACTAGGTTGCTGCGCCCATCCGACCAGTACAAATCTAGCAGTAGATAAATTGCAGTGGCTACATCTTGTAAGTCGCACCAGTCGATTTGAGGATGTACCTAGGTCACAAAAGCACTGACGTACTATTGCGGGAACGTAAGAGCTAAGCCCAATTGCGT\n"
     ]
    }
   ],
   "source": [
    "seq =''\n",
    "seqlist = [] #Create an empty list to store values - both labels and sequences \n",
    "with open(\"exercises/rosalind_lcsm.txt\") as file_in:\n",
    "    for line in file_in:\n",
    "        if line[0] == \">\":\n",
    "            line = line.lstrip('>')\n",
    "            line = line.rstrip('\\n')\n",
    "            seqlist.append(seq)\n",
    "            seqlist.append(line)\n",
    "            seq = \"\"\n",
    "        else:\n",
    "            line = line.rstrip('\\n')\n",
    "            seq += line\n",
    "    seqlist.append(seq) #When it runs out of lines, it'll append the last sequence\n",
    "seqlist.pop(0) #Remove first empty space\n",
    "\n",
    "sequences = [] #Collect all sequences\n",
    "for i in range(1, len(seqlist), 2):\n",
    "    sequences.append(seqlist[i])\n",
    "\n",
    "shortest = min(sequences, key=len) \n",
    "#Let's take out the shortest sequence for analysis since the longest common substring must be from this sequence\n",
    "sequences.remove(shortest)\n",
    "\n",
    "#Generate all possible LCS candidates from the shortest string\n",
    "stringcandidates = []\n",
    "for i in range(0, len(shortest)):\n",
    "    for j in range(len(shortest), 0, -1): \n",
    "        if j-i > 1:\n",
    "            stringcandidates.append(shortest[i:j])\n",
    "stringcandidates.sort(key=len, reverse=True) #Order by length so we search for the longest first\n",
    "\n",
    "#For every candidate LCS, search every string with a window corresponding to LCS length.\n",
    "for i in range(0, len(stringcandidates)):\n",
    "    stringlength = len(stringcandidates[i])\n",
    "    counter = 0\n",
    "    for j in range(0, len(sequences)):\n",
    "        stringcounter = 0\n",
    "        for k in range(0, len(sequences[j])):\n",
    "            if sequences[j][k:k+stringlength] == stringcandidates[i]:\n",
    "                stringcounter += 1\n",
    "                break #If we have a match, move on to next string.\n",
    "        if stringcounter == 0:\n",
    "            break #If we had no match, discard the LCS.\n",
    "        else: \n",
    "            counter += stringcounter\n",
    "    if counter == len(sequences): #If the number of matches in total is equal to the number of strings, we have a successful candidate\n",
    "        print(stringcandidates[i])\n",
    "        break #Since we search from the longest to shortest, the first successful candidate is by definition the longest of the bunch\n",
    "        #We terminate the for loop here to save time and power"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 21 - Enumerating Gene Orders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "720\n",
      "1 2 3 4 5 6\n",
      "1 2 3 4 6 5\n",
      "1 2 3 5 4 6\n",
      "1 2 3 5 6 4\n",
      "1 2 3 6 4 5\n",
      "1 2 3 6 5 4\n",
      "1 2 4 3 5 6\n",
      "1 2 4 3 6 5\n",
      "1 2 4 5 3 6\n",
      "1 2 4 5 6 3\n",
      "1 2 4 6 3 5\n",
      "1 2 4 6 5 3\n",
      "1 2 5 3 4 6\n",
      "1 2 5 3 6 4\n",
      "1 2 5 4 3 6\n",
      "1 2 5 4 6 3\n",
      "1 2 5 6 3 4\n",
      "1 2 5 6 4 3\n",
      "1 2 6 3 4 5\n",
      "1 2 6 3 5 4\n",
      "1 2 6 4 3 5\n",
      "1 2 6 4 5 3\n",
      "1 2 6 5 3 4\n",
      "1 2 6 5 4 3\n",
      "1 3 2 4 5 6\n",
      "1 3 2 4 6 5\n",
      "1 3 2 5 4 6\n",
      "1 3 2 5 6 4\n",
      "1 3 2 6 4 5\n",
      "1 3 2 6 5 4\n",
      "1 3 4 2 5 6\n",
      "1 3 4 2 6 5\n",
      "1 3 4 5 2 6\n",
      "1 3 4 5 6 2\n",
      "1 3 4 6 2 5\n",
      "1 3 4 6 5 2\n",
      "1 3 5 2 4 6\n",
      "1 3 5 2 6 4\n",
      "1 3 5 4 2 6\n",
      "1 3 5 4 6 2\n",
      "1 3 5 6 2 4\n",
      "1 3 5 6 4 2\n",
      "1 3 6 2 4 5\n",
      "1 3 6 2 5 4\n",
      "1 3 6 4 2 5\n",
      "1 3 6 4 5 2\n",
      "1 3 6 5 2 4\n",
      "1 3 6 5 4 2\n",
      "1 4 2 3 5 6\n",
      "1 4 2 3 6 5\n",
      "1 4 2 5 3 6\n",
      "1 4 2 5 6 3\n",
      "1 4 2 6 3 5\n",
      "1 4 2 6 5 3\n",
      "1 4 3 2 5 6\n",
      "1 4 3 2 6 5\n",
      "1 4 3 5 2 6\n",
      "1 4 3 5 6 2\n",
      "1 4 3 6 2 5\n",
      "1 4 3 6 5 2\n",
      "1 4 5 2 3 6\n",
      "1 4 5 2 6 3\n",
      "1 4 5 3 2 6\n",
      "1 4 5 3 6 2\n",
      "1 4 5 6 2 3\n",
      "1 4 5 6 3 2\n",
      "1 4 6 2 3 5\n",
      "1 4 6 2 5 3\n",
      "1 4 6 3 2 5\n",
      "1 4 6 3 5 2\n",
      "1 4 6 5 2 3\n",
      "1 4 6 5 3 2\n",
      "1 5 2 3 4 6\n",
      "1 5 2 3 6 4\n",
      "1 5 2 4 3 6\n",
      "1 5 2 4 6 3\n",
      "1 5 2 6 3 4\n",
      "1 5 2 6 4 3\n",
      "1 5 3 2 4 6\n",
      "1 5 3 2 6 4\n",
      "1 5 3 4 2 6\n",
      "1 5 3 4 6 2\n",
      "1 5 3 6 2 4\n",
      "1 5 3 6 4 2\n",
      "1 5 4 2 3 6\n",
      "1 5 4 2 6 3\n",
      "1 5 4 3 2 6\n",
      "1 5 4 3 6 2\n",
      "1 5 4 6 2 3\n",
      "1 5 4 6 3 2\n",
      "1 5 6 2 3 4\n",
      "1 5 6 2 4 3\n",
      "1 5 6 3 2 4\n",
      "1 5 6 3 4 2\n",
      "1 5 6 4 2 3\n",
      "1 5 6 4 3 2\n",
      "1 6 2 3 4 5\n",
      "1 6 2 3 5 4\n",
      "1 6 2 4 3 5\n",
      "1 6 2 4 5 3\n",
      "1 6 2 5 3 4\n",
      "1 6 2 5 4 3\n",
      "1 6 3 2 4 5\n",
      "1 6 3 2 5 4\n",
      "1 6 3 4 2 5\n",
      "1 6 3 4 5 2\n",
      "1 6 3 5 2 4\n",
      "1 6 3 5 4 2\n",
      "1 6 4 2 3 5\n",
      "1 6 4 2 5 3\n",
      "1 6 4 3 2 5\n",
      "1 6 4 3 5 2\n",
      "1 6 4 5 2 3\n",
      "1 6 4 5 3 2\n",
      "1 6 5 2 3 4\n",
      "1 6 5 2 4 3\n",
      "1 6 5 3 2 4\n",
      "1 6 5 3 4 2\n",
      "1 6 5 4 2 3\n",
      "1 6 5 4 3 2\n",
      "2 1 3 4 5 6\n",
      "2 1 3 4 6 5\n",
      "2 1 3 5 4 6\n",
      "2 1 3 5 6 4\n",
      "2 1 3 6 4 5\n",
      "2 1 3 6 5 4\n",
      "2 1 4 3 5 6\n",
      "2 1 4 3 6 5\n",
      "2 1 4 5 3 6\n",
      "2 1 4 5 6 3\n",
      "2 1 4 6 3 5\n",
      "2 1 4 6 5 3\n",
      "2 1 5 3 4 6\n",
      "2 1 5 3 6 4\n",
      "2 1 5 4 3 6\n",
      "2 1 5 4 6 3\n",
      "2 1 5 6 3 4\n",
      "2 1 5 6 4 3\n",
      "2 1 6 3 4 5\n",
      "2 1 6 3 5 4\n",
      "2 1 6 4 3 5\n",
      "2 1 6 4 5 3\n",
      "2 1 6 5 3 4\n",
      "2 1 6 5 4 3\n",
      "2 3 1 4 5 6\n",
      "2 3 1 4 6 5\n",
      "2 3 1 5 4 6\n",
      "2 3 1 5 6 4\n",
      "2 3 1 6 4 5\n",
      "2 3 1 6 5 4\n",
      "2 3 4 1 5 6\n",
      "2 3 4 1 6 5\n",
      "2 3 4 5 1 6\n",
      "2 3 4 5 6 1\n",
      "2 3 4 6 1 5\n",
      "2 3 4 6 5 1\n",
      "2 3 5 1 4 6\n",
      "2 3 5 1 6 4\n",
      "2 3 5 4 1 6\n",
      "2 3 5 4 6 1\n",
      "2 3 5 6 1 4\n",
      "2 3 5 6 4 1\n",
      "2 3 6 1 4 5\n",
      "2 3 6 1 5 4\n",
      "2 3 6 4 1 5\n",
      "2 3 6 4 5 1\n",
      "2 3 6 5 1 4\n",
      "2 3 6 5 4 1\n",
      "2 4 1 3 5 6\n",
      "2 4 1 3 6 5\n",
      "2 4 1 5 3 6\n",
      "2 4 1 5 6 3\n",
      "2 4 1 6 3 5\n",
      "2 4 1 6 5 3\n",
      "2 4 3 1 5 6\n",
      "2 4 3 1 6 5\n",
      "2 4 3 5 1 6\n",
      "2 4 3 5 6 1\n",
      "2 4 3 6 1 5\n",
      "2 4 3 6 5 1\n",
      "2 4 5 1 3 6\n",
      "2 4 5 1 6 3\n",
      "2 4 5 3 1 6\n",
      "2 4 5 3 6 1\n",
      "2 4 5 6 1 3\n",
      "2 4 5 6 3 1\n",
      "2 4 6 1 3 5\n",
      "2 4 6 1 5 3\n",
      "2 4 6 3 1 5\n",
      "2 4 6 3 5 1\n",
      "2 4 6 5 1 3\n",
      "2 4 6 5 3 1\n",
      "2 5 1 3 4 6\n",
      "2 5 1 3 6 4\n",
      "2 5 1 4 3 6\n",
      "2 5 1 4 6 3\n",
      "2 5 1 6 3 4\n",
      "2 5 1 6 4 3\n",
      "2 5 3 1 4 6\n",
      "2 5 3 1 6 4\n",
      "2 5 3 4 1 6\n",
      "2 5 3 4 6 1\n",
      "2 5 3 6 1 4\n",
      "2 5 3 6 4 1\n",
      "2 5 4 1 3 6\n",
      "2 5 4 1 6 3\n",
      "2 5 4 3 1 6\n",
      "2 5 4 3 6 1\n",
      "2 5 4 6 1 3\n",
      "2 5 4 6 3 1\n",
      "2 5 6 1 3 4\n",
      "2 5 6 1 4 3\n",
      "2 5 6 3 1 4\n",
      "2 5 6 3 4 1\n",
      "2 5 6 4 1 3\n",
      "2 5 6 4 3 1\n",
      "2 6 1 3 4 5\n",
      "2 6 1 3 5 4\n",
      "2 6 1 4 3 5\n",
      "2 6 1 4 5 3\n",
      "2 6 1 5 3 4\n",
      "2 6 1 5 4 3\n",
      "2 6 3 1 4 5\n",
      "2 6 3 1 5 4\n",
      "2 6 3 4 1 5\n",
      "2 6 3 4 5 1\n",
      "2 6 3 5 1 4\n",
      "2 6 3 5 4 1\n",
      "2 6 4 1 3 5\n",
      "2 6 4 1 5 3\n",
      "2 6 4 3 1 5\n",
      "2 6 4 3 5 1\n",
      "2 6 4 5 1 3\n",
      "2 6 4 5 3 1\n",
      "2 6 5 1 3 4\n",
      "2 6 5 1 4 3\n",
      "2 6 5 3 1 4\n",
      "2 6 5 3 4 1\n",
      "2 6 5 4 1 3\n",
      "2 6 5 4 3 1\n",
      "3 1 2 4 5 6\n",
      "3 1 2 4 6 5\n",
      "3 1 2 5 4 6\n",
      "3 1 2 5 6 4\n",
      "3 1 2 6 4 5\n",
      "3 1 2 6 5 4\n",
      "3 1 4 2 5 6\n",
      "3 1 4 2 6 5\n",
      "3 1 4 5 2 6\n",
      "3 1 4 5 6 2\n",
      "3 1 4 6 2 5\n",
      "3 1 4 6 5 2\n",
      "3 1 5 2 4 6\n",
      "3 1 5 2 6 4\n",
      "3 1 5 4 2 6\n",
      "3 1 5 4 6 2\n",
      "3 1 5 6 2 4\n",
      "3 1 5 6 4 2\n",
      "3 1 6 2 4 5\n",
      "3 1 6 2 5 4\n",
      "3 1 6 4 2 5\n",
      "3 1 6 4 5 2\n",
      "3 1 6 5 2 4\n",
      "3 1 6 5 4 2\n",
      "3 2 1 4 5 6\n",
      "3 2 1 4 6 5\n",
      "3 2 1 5 4 6\n",
      "3 2 1 5 6 4\n",
      "3 2 1 6 4 5\n",
      "3 2 1 6 5 4\n",
      "3 2 4 1 5 6\n",
      "3 2 4 1 6 5\n",
      "3 2 4 5 1 6\n",
      "3 2 4 5 6 1\n",
      "3 2 4 6 1 5\n",
      "3 2 4 6 5 1\n",
      "3 2 5 1 4 6\n",
      "3 2 5 1 6 4\n",
      "3 2 5 4 1 6\n",
      "3 2 5 4 6 1\n",
      "3 2 5 6 1 4\n",
      "3 2 5 6 4 1\n",
      "3 2 6 1 4 5\n",
      "3 2 6 1 5 4\n",
      "3 2 6 4 1 5\n",
      "3 2 6 4 5 1\n",
      "3 2 6 5 1 4\n",
      "3 2 6 5 4 1\n",
      "3 4 1 2 5 6\n",
      "3 4 1 2 6 5\n",
      "3 4 1 5 2 6\n",
      "3 4 1 5 6 2\n",
      "3 4 1 6 2 5\n",
      "3 4 1 6 5 2\n",
      "3 4 2 1 5 6\n",
      "3 4 2 1 6 5\n",
      "3 4 2 5 1 6\n",
      "3 4 2 5 6 1\n",
      "3 4 2 6 1 5\n",
      "3 4 2 6 5 1\n",
      "3 4 5 1 2 6\n",
      "3 4 5 1 6 2\n",
      "3 4 5 2 1 6\n",
      "3 4 5 2 6 1\n",
      "3 4 5 6 1 2\n",
      "3 4 5 6 2 1\n",
      "3 4 6 1 2 5\n",
      "3 4 6 1 5 2\n",
      "3 4 6 2 1 5\n",
      "3 4 6 2 5 1\n",
      "3 4 6 5 1 2\n",
      "3 4 6 5 2 1\n",
      "3 5 1 2 4 6\n",
      "3 5 1 2 6 4\n",
      "3 5 1 4 2 6\n",
      "3 5 1 4 6 2\n",
      "3 5 1 6 2 4\n",
      "3 5 1 6 4 2\n",
      "3 5 2 1 4 6\n",
      "3 5 2 1 6 4\n",
      "3 5 2 4 1 6\n",
      "3 5 2 4 6 1\n",
      "3 5 2 6 1 4\n",
      "3 5 2 6 4 1\n",
      "3 5 4 1 2 6\n",
      "3 5 4 1 6 2\n",
      "3 5 4 2 1 6\n",
      "3 5 4 2 6 1\n",
      "3 5 4 6 1 2\n",
      "3 5 4 6 2 1\n",
      "3 5 6 1 2 4\n",
      "3 5 6 1 4 2\n",
      "3 5 6 2 1 4\n",
      "3 5 6 2 4 1\n",
      "3 5 6 4 1 2\n",
      "3 5 6 4 2 1\n",
      "3 6 1 2 4 5\n",
      "3 6 1 2 5 4\n",
      "3 6 1 4 2 5\n",
      "3 6 1 4 5 2\n",
      "3 6 1 5 2 4\n",
      "3 6 1 5 4 2\n",
      "3 6 2 1 4 5\n",
      "3 6 2 1 5 4\n",
      "3 6 2 4 1 5\n",
      "3 6 2 4 5 1\n",
      "3 6 2 5 1 4\n",
      "3 6 2 5 4 1\n",
      "3 6 4 1 2 5\n",
      "3 6 4 1 5 2\n",
      "3 6 4 2 1 5\n",
      "3 6 4 2 5 1\n",
      "3 6 4 5 1 2\n",
      "3 6 4 5 2 1\n",
      "3 6 5 1 2 4\n",
      "3 6 5 1 4 2\n",
      "3 6 5 2 1 4\n",
      "3 6 5 2 4 1\n",
      "3 6 5 4 1 2\n",
      "3 6 5 4 2 1\n",
      "4 1 2 3 5 6\n",
      "4 1 2 3 6 5\n",
      "4 1 2 5 3 6\n",
      "4 1 2 5 6 3\n",
      "4 1 2 6 3 5\n",
      "4 1 2 6 5 3\n",
      "4 1 3 2 5 6\n",
      "4 1 3 2 6 5\n",
      "4 1 3 5 2 6\n",
      "4 1 3 5 6 2\n",
      "4 1 3 6 2 5\n",
      "4 1 3 6 5 2\n",
      "4 1 5 2 3 6\n",
      "4 1 5 2 6 3\n",
      "4 1 5 3 2 6\n",
      "4 1 5 3 6 2\n",
      "4 1 5 6 2 3\n",
      "4 1 5 6 3 2\n",
      "4 1 6 2 3 5\n",
      "4 1 6 2 5 3\n",
      "4 1 6 3 2 5\n",
      "4 1 6 3 5 2\n",
      "4 1 6 5 2 3\n",
      "4 1 6 5 3 2\n",
      "4 2 1 3 5 6\n",
      "4 2 1 3 6 5\n",
      "4 2 1 5 3 6\n",
      "4 2 1 5 6 3\n",
      "4 2 1 6 3 5\n",
      "4 2 1 6 5 3\n",
      "4 2 3 1 5 6\n",
      "4 2 3 1 6 5\n",
      "4 2 3 5 1 6\n",
      "4 2 3 5 6 1\n",
      "4 2 3 6 1 5\n",
      "4 2 3 6 5 1\n",
      "4 2 5 1 3 6\n",
      "4 2 5 1 6 3\n",
      "4 2 5 3 1 6\n",
      "4 2 5 3 6 1\n",
      "4 2 5 6 1 3\n",
      "4 2 5 6 3 1\n",
      "4 2 6 1 3 5\n",
      "4 2 6 1 5 3\n",
      "4 2 6 3 1 5\n",
      "4 2 6 3 5 1\n",
      "4 2 6 5 1 3\n",
      "4 2 6 5 3 1\n",
      "4 3 1 2 5 6\n",
      "4 3 1 2 6 5\n",
      "4 3 1 5 2 6\n",
      "4 3 1 5 6 2\n",
      "4 3 1 6 2 5\n",
      "4 3 1 6 5 2\n",
      "4 3 2 1 5 6\n",
      "4 3 2 1 6 5\n",
      "4 3 2 5 1 6\n",
      "4 3 2 5 6 1\n",
      "4 3 2 6 1 5\n",
      "4 3 2 6 5 1\n",
      "4 3 5 1 2 6\n",
      "4 3 5 1 6 2\n",
      "4 3 5 2 1 6\n",
      "4 3 5 2 6 1\n",
      "4 3 5 6 1 2\n",
      "4 3 5 6 2 1\n",
      "4 3 6 1 2 5\n",
      "4 3 6 1 5 2\n",
      "4 3 6 2 1 5\n",
      "4 3 6 2 5 1\n",
      "4 3 6 5 1 2\n",
      "4 3 6 5 2 1\n",
      "4 5 1 2 3 6\n",
      "4 5 1 2 6 3\n",
      "4 5 1 3 2 6\n",
      "4 5 1 3 6 2\n",
      "4 5 1 6 2 3\n",
      "4 5 1 6 3 2\n",
      "4 5 2 1 3 6\n",
      "4 5 2 1 6 3\n",
      "4 5 2 3 1 6\n",
      "4 5 2 3 6 1\n",
      "4 5 2 6 1 3\n",
      "4 5 2 6 3 1\n",
      "4 5 3 1 2 6\n",
      "4 5 3 1 6 2\n",
      "4 5 3 2 1 6\n",
      "4 5 3 2 6 1\n",
      "4 5 3 6 1 2\n",
      "4 5 3 6 2 1\n",
      "4 5 6 1 2 3\n",
      "4 5 6 1 3 2\n",
      "4 5 6 2 1 3\n",
      "4 5 6 2 3 1\n",
      "4 5 6 3 1 2\n",
      "4 5 6 3 2 1\n",
      "4 6 1 2 3 5\n",
      "4 6 1 2 5 3\n",
      "4 6 1 3 2 5\n",
      "4 6 1 3 5 2\n",
      "4 6 1 5 2 3\n",
      "4 6 1 5 3 2\n",
      "4 6 2 1 3 5\n",
      "4 6 2 1 5 3\n",
      "4 6 2 3 1 5\n",
      "4 6 2 3 5 1\n",
      "4 6 2 5 1 3\n",
      "4 6 2 5 3 1\n",
      "4 6 3 1 2 5\n",
      "4 6 3 1 5 2\n",
      "4 6 3 2 1 5\n",
      "4 6 3 2 5 1\n",
      "4 6 3 5 1 2\n",
      "4 6 3 5 2 1\n",
      "4 6 5 1 2 3\n",
      "4 6 5 1 3 2\n",
      "4 6 5 2 1 3\n",
      "4 6 5 2 3 1\n",
      "4 6 5 3 1 2\n",
      "4 6 5 3 2 1\n",
      "5 1 2 3 4 6\n",
      "5 1 2 3 6 4\n",
      "5 1 2 4 3 6\n",
      "5 1 2 4 6 3\n",
      "5 1 2 6 3 4\n",
      "5 1 2 6 4 3\n",
      "5 1 3 2 4 6\n",
      "5 1 3 2 6 4\n",
      "5 1 3 4 2 6\n",
      "5 1 3 4 6 2\n",
      "5 1 3 6 2 4\n",
      "5 1 3 6 4 2\n",
      "5 1 4 2 3 6\n",
      "5 1 4 2 6 3\n",
      "5 1 4 3 2 6\n",
      "5 1 4 3 6 2\n",
      "5 1 4 6 2 3\n",
      "5 1 4 6 3 2\n",
      "5 1 6 2 3 4\n",
      "5 1 6 2 4 3\n",
      "5 1 6 3 2 4\n",
      "5 1 6 3 4 2\n",
      "5 1 6 4 2 3\n",
      "5 1 6 4 3 2\n",
      "5 2 1 3 4 6\n",
      "5 2 1 3 6 4\n",
      "5 2 1 4 3 6\n",
      "5 2 1 4 6 3\n",
      "5 2 1 6 3 4\n",
      "5 2 1 6 4 3\n",
      "5 2 3 1 4 6\n",
      "5 2 3 1 6 4\n",
      "5 2 3 4 1 6\n",
      "5 2 3 4 6 1\n",
      "5 2 3 6 1 4\n",
      "5 2 3 6 4 1\n",
      "5 2 4 1 3 6\n",
      "5 2 4 1 6 3\n",
      "5 2 4 3 1 6\n",
      "5 2 4 3 6 1\n",
      "5 2 4 6 1 3\n",
      "5 2 4 6 3 1\n",
      "5 2 6 1 3 4\n",
      "5 2 6 1 4 3\n",
      "5 2 6 3 1 4\n",
      "5 2 6 3 4 1\n",
      "5 2 6 4 1 3\n",
      "5 2 6 4 3 1\n",
      "5 3 1 2 4 6\n",
      "5 3 1 2 6 4\n",
      "5 3 1 4 2 6\n",
      "5 3 1 4 6 2\n",
      "5 3 1 6 2 4\n",
      "5 3 1 6 4 2\n",
      "5 3 2 1 4 6\n",
      "5 3 2 1 6 4\n",
      "5 3 2 4 1 6\n",
      "5 3 2 4 6 1\n",
      "5 3 2 6 1 4\n",
      "5 3 2 6 4 1\n",
      "5 3 4 1 2 6\n",
      "5 3 4 1 6 2\n",
      "5 3 4 2 1 6\n",
      "5 3 4 2 6 1\n",
      "5 3 4 6 1 2\n",
      "5 3 4 6 2 1\n",
      "5 3 6 1 2 4\n",
      "5 3 6 1 4 2\n",
      "5 3 6 2 1 4\n",
      "5 3 6 2 4 1\n",
      "5 3 6 4 1 2\n",
      "5 3 6 4 2 1\n",
      "5 4 1 2 3 6\n",
      "5 4 1 2 6 3\n",
      "5 4 1 3 2 6\n",
      "5 4 1 3 6 2\n",
      "5 4 1 6 2 3\n",
      "5 4 1 6 3 2\n",
      "5 4 2 1 3 6\n",
      "5 4 2 1 6 3\n",
      "5 4 2 3 1 6\n",
      "5 4 2 3 6 1\n",
      "5 4 2 6 1 3\n",
      "5 4 2 6 3 1\n",
      "5 4 3 1 2 6\n",
      "5 4 3 1 6 2\n",
      "5 4 3 2 1 6\n",
      "5 4 3 2 6 1\n",
      "5 4 3 6 1 2\n",
      "5 4 3 6 2 1\n",
      "5 4 6 1 2 3\n",
      "5 4 6 1 3 2\n",
      "5 4 6 2 1 3\n",
      "5 4 6 2 3 1\n",
      "5 4 6 3 1 2\n",
      "5 4 6 3 2 1\n",
      "5 6 1 2 3 4\n",
      "5 6 1 2 4 3\n",
      "5 6 1 3 2 4\n",
      "5 6 1 3 4 2\n",
      "5 6 1 4 2 3\n",
      "5 6 1 4 3 2\n",
      "5 6 2 1 3 4\n",
      "5 6 2 1 4 3\n",
      "5 6 2 3 1 4\n",
      "5 6 2 3 4 1\n",
      "5 6 2 4 1 3\n",
      "5 6 2 4 3 1\n",
      "5 6 3 1 2 4\n",
      "5 6 3 1 4 2\n",
      "5 6 3 2 1 4\n",
      "5 6 3 2 4 1\n",
      "5 6 3 4 1 2\n",
      "5 6 3 4 2 1\n",
      "5 6 4 1 2 3\n",
      "5 6 4 1 3 2\n",
      "5 6 4 2 1 3\n",
      "5 6 4 2 3 1\n",
      "5 6 4 3 1 2\n",
      "5 6 4 3 2 1\n",
      "6 1 2 3 4 5\n",
      "6 1 2 3 5 4\n",
      "6 1 2 4 3 5\n",
      "6 1 2 4 5 3\n",
      "6 1 2 5 3 4\n",
      "6 1 2 5 4 3\n",
      "6 1 3 2 4 5\n",
      "6 1 3 2 5 4\n",
      "6 1 3 4 2 5\n",
      "6 1 3 4 5 2\n",
      "6 1 3 5 2 4\n",
      "6 1 3 5 4 2\n",
      "6 1 4 2 3 5\n",
      "6 1 4 2 5 3\n",
      "6 1 4 3 2 5\n",
      "6 1 4 3 5 2\n",
      "6 1 4 5 2 3\n",
      "6 1 4 5 3 2\n",
      "6 1 5 2 3 4\n",
      "6 1 5 2 4 3\n",
      "6 1 5 3 2 4\n",
      "6 1 5 3 4 2\n",
      "6 1 5 4 2 3\n",
      "6 1 5 4 3 2\n",
      "6 2 1 3 4 5\n",
      "6 2 1 3 5 4\n",
      "6 2 1 4 3 5\n",
      "6 2 1 4 5 3\n",
      "6 2 1 5 3 4\n",
      "6 2 1 5 4 3\n",
      "6 2 3 1 4 5\n",
      "6 2 3 1 5 4\n",
      "6 2 3 4 1 5\n",
      "6 2 3 4 5 1\n",
      "6 2 3 5 1 4\n",
      "6 2 3 5 4 1\n",
      "6 2 4 1 3 5\n",
      "6 2 4 1 5 3\n",
      "6 2 4 3 1 5\n",
      "6 2 4 3 5 1\n",
      "6 2 4 5 1 3\n",
      "6 2 4 5 3 1\n",
      "6 2 5 1 3 4\n",
      "6 2 5 1 4 3\n",
      "6 2 5 3 1 4\n",
      "6 2 5 3 4 1\n",
      "6 2 5 4 1 3\n",
      "6 2 5 4 3 1\n",
      "6 3 1 2 4 5\n",
      "6 3 1 2 5 4\n",
      "6 3 1 4 2 5\n",
      "6 3 1 4 5 2\n",
      "6 3 1 5 2 4\n",
      "6 3 1 5 4 2\n",
      "6 3 2 1 4 5\n",
      "6 3 2 1 5 4\n",
      "6 3 2 4 1 5\n",
      "6 3 2 4 5 1\n",
      "6 3 2 5 1 4\n",
      "6 3 2 5 4 1\n",
      "6 3 4 1 2 5\n",
      "6 3 4 1 5 2\n",
      "6 3 4 2 1 5\n",
      "6 3 4 2 5 1\n",
      "6 3 4 5 1 2\n",
      "6 3 4 5 2 1\n",
      "6 3 5 1 2 4\n",
      "6 3 5 1 4 2\n",
      "6 3 5 2 1 4\n",
      "6 3 5 2 4 1\n",
      "6 3 5 4 1 2\n",
      "6 3 5 4 2 1\n",
      "6 4 1 2 3 5\n",
      "6 4 1 2 5 3\n",
      "6 4 1 3 2 5\n",
      "6 4 1 3 5 2\n",
      "6 4 1 5 2 3\n",
      "6 4 1 5 3 2\n",
      "6 4 2 1 3 5\n",
      "6 4 2 1 5 3\n",
      "6 4 2 3 1 5\n",
      "6 4 2 3 5 1\n",
      "6 4 2 5 1 3\n",
      "6 4 2 5 3 1\n",
      "6 4 3 1 2 5\n",
      "6 4 3 1 5 2\n",
      "6 4 3 2 1 5\n",
      "6 4 3 2 5 1\n",
      "6 4 3 5 1 2\n",
      "6 4 3 5 2 1\n",
      "6 4 5 1 2 3\n",
      "6 4 5 1 3 2\n",
      "6 4 5 2 1 3\n",
      "6 4 5 2 3 1\n",
      "6 4 5 3 1 2\n",
      "6 4 5 3 2 1\n",
      "6 5 1 2 3 4\n",
      "6 5 1 2 4 3\n",
      "6 5 1 3 2 4\n",
      "6 5 1 3 4 2\n",
      "6 5 1 4 2 3\n",
      "6 5 1 4 3 2\n",
      "6 5 2 1 3 4\n",
      "6 5 2 1 4 3\n",
      "6 5 2 3 1 4\n",
      "6 5 2 3 4 1\n",
      "6 5 2 4 1 3\n",
      "6 5 2 4 3 1\n",
      "6 5 3 1 2 4\n",
      "6 5 3 1 4 2\n",
      "6 5 3 2 1 4\n",
      "6 5 3 2 4 1\n",
      "6 5 3 4 1 2\n",
      "6 5 3 4 2 1\n",
      "6 5 4 1 2 3\n",
      "6 5 4 1 3 2\n",
      "6 5 4 2 1 3\n",
      "6 5 4 2 3 1\n",
      "6 5 4 3 1 2\n",
      "6 5 4 3 2 1\n"
     ]
    }
   ],
   "source": [
    "#from rosalind_perm.txt\n",
    "\n",
    "import math\n",
    "import itertools\n",
    "\n",
    "n = 6\n",
    "\n",
    "print(math.factorial(n))\n",
    "\n",
    "numbers = []\n",
    "for i in range(1, n+1):\n",
    "    numbers.append(i)\n",
    "\n",
    "perm = list(itertools.permutations(numbers))\n",
    "\n",
    "for i in range(0, len(perm)):\n",
    "    perm[i] = str(perm[i])\n",
    "    perm[i] = perm[i].replace(\"(\", \"\")\n",
    "    perm[i] = perm[i].replace(\",\", \"\")\n",
    "    perm[i] = perm[i].replace(\")\", \"\")\n",
    "    print(perm[i])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
